
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../../tools/_segmentation/">
      
      
        <link rel="next" href="../../tutorials/mapping_sc_bulk/">
      
      
      <link rel="icon" href="../../assets/favicon_temp.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.5.49">
    
    
      
        <title>plot - CytoBulk's Docs</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.6f8fc17f.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="../../css/ansi-colours.css">
    
      <link rel="stylesheet" href="../../css/jupyter-cells.css">
    
      <link rel="stylesheet" href="../../css/pandas-dataframe.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#plots_plot" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="CytoBulk&#39;s Docs" class="md-header__button md-logo" aria-label="CytoBulk's Docs" data-md-component="logo">
      
  <img src="../../assets/logo_temp.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            CytoBulk's Docs
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              plot
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m14.3 16-.7-2h-3.2l-.7 2H7.8L11 7h2l3.2 9zM20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12zm-9.15 3.96h2.3L12 9z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_2" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 7a5 5 0 0 1 5 5 5 5 0 0 1-5 5 5 5 0 0 1-5-5 5 5 0 0 1 5-5m0 2a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0-7 2.39 3.42C13.65 5.15 12.84 5 12 5s-1.65.15-2.39.42zM3.34 7l4.16-.35A7.2 7.2 0 0 0 5.94 8.5c-.44.74-.69 1.5-.83 2.29zm.02 10 1.76-3.77a7.131 7.131 0 0 0 2.38 4.14zM20.65 7l-1.77 3.79a7.02 7.02 0 0 0-2.38-4.15zm-.01 10-4.14.36c.59-.51 1.12-1.14 1.54-1.86.42-.73.69-1.5.83-2.29zM12 22l-2.41-3.44c.74.27 1.55.44 2.41.44.82 0 1.63-.17 2.37-.44z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_2">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3zm3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95zm-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://kristaxying.github.io/CytoBulk/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    CytoBulk
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="CytoBulk&#39;s Docs" class="md-nav__button md-logo" aria-label="CytoBulk's Docs" data-md-component="logo">
      
  <img src="../../assets/logo_temp.png" alt="logo">

    </a>
    CytoBulk's Docs
  </label>
  
    <div class="md-nav__source">
      <a href="https://kristaxying.github.io/CytoBulk/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    CytoBulk
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../Install/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Install
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" checked>
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    API References
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            API References
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tools/_deconv/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    deconv
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tools/_mapping/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    mapping
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tools/_segmentation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    segmentation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    plot
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    plot
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#cytobulk.plots._plot.plot_celltype_fraction_pie" class="md-nav__link">
    <span class="md-ellipsis">
      plot_celltype_fraction_pie
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#cytobulk.plots._plot.plot_celltype_fraction_heatmap" class="md-nav__link">
    <span class="md-ellipsis">
      plot_celltype_fraction_heatmap
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#cytobulk.plots._plot.plot_paired_violin" class="md-nav__link">
    <span class="md-ellipsis">
      plot_paired_violin
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#cytobulk.plots._plot.plot_reconstruction" class="md-nav__link">
    <span class="md-ellipsis">
      plot_reconstruction
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#cytobulk.plots._plot.plot_gene_similarity" class="md-nav__link">
    <span class="md-ellipsis">
      plot_gene_similarity
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#cytobulk.plots._plot.plot_he_cell_type" class="md-nav__link">
    <span class="md-ellipsis">
      plot_he_cell_type
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Examples
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Examples
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tutorials/mapping_sc_bulk/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Bulk deconvolution and mapping with scRNA-seq
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tutorials/mapping_sc_st/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ST deconvolution and mapping with scRNA-seq
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tutorials/mapping_he_sc/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    H&E cell prediction and integration with scRNA-seq
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../References/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    References
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#cytobulk.plots._plot.plot_celltype_fraction_pie" class="md-nav__link">
    <span class="md-ellipsis">
      plot_celltype_fraction_pie
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#cytobulk.plots._plot.plot_celltype_fraction_heatmap" class="md-nav__link">
    <span class="md-ellipsis">
      plot_celltype_fraction_heatmap
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#cytobulk.plots._plot.plot_paired_violin" class="md-nav__link">
    <span class="md-ellipsis">
      plot_paired_violin
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#cytobulk.plots._plot.plot_reconstruction" class="md-nav__link">
    <span class="md-ellipsis">
      plot_reconstruction
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#cytobulk.plots._plot.plot_gene_similarity" class="md-nav__link">
    <span class="md-ellipsis">
      plot_gene_similarity
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#cytobulk.plots._plot.plot_he_cell_type" class="md-nav__link">
    <span class="md-ellipsis">
      plot_he_cell_type
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


<h1 id="plots_plot">plots._plot</h1>
<hr />
<p><div class="highlight"><pre><span></span><code><span class="n">plot_celltype_fraction_pie</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span>
                               <span class="n">scale_facter_x</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
                               <span class="n">scale_factor_y</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
                               <span class="n">r</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
                               <span class="n">out_dir</span><span class="o">=</span><span class="s1">&#39;/plot&#39;</span><span class="p">,</span>
                               <span class="n">project_name</span><span class="o">=</span><span class="s1">&#39;project&#39;</span><span class="p">,</span>
                               <span class="n">color_list</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                               <span class="n">color_map</span><span class="o">=</span><span class="s2">&quot;Spectral&quot;</span><span class="p">,</span>
                               <span class="n">rotation_angle</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                               <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span>
</code></pre></div>
</p>


<div class="doc doc-object doc-function">



<a id="cytobulk.plots._plot.plot_celltype_fraction_pie"></a>
    <div class="doc doc-contents first">

        <p>Plot cell type fraction as pie charts at spatial coordinates.</p>
<p>This function generates pie charts representing the cell type fraction at each spatial location (from the 'spatial' data in <code>adata</code>) and plots them on a 2D scatter plot. Each pie chart is positioned based on the spatial coordinates and reflects the relative abundance of each cell type at that spot. The resulting figure is saved as an SVG file.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>adata</code>
            </td>
            <td>
                  <code><span title="anndata.AnnData">AnnData</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>An :class:<code>~anndata.AnnData</code> object containing spatial transcriptomics data.
The spatial coordinates should be stored in <code>adata.obsm['spatial']</code>, and the deconvolution results (cell type fractions) should be stored in <code>adata.uns['deconv']</code>.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>scale_facter_x</code>
            </td>
            <td>
                  <code>float, optional (default: 1)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A scaling factor to apply to the X-axis coordinates of the spatial data.</p>
              </div>
            </td>
            <td>
                  <code>1</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>scale_factor_y</code>
            </td>
            <td>
                  <code>float, optional (default: 1)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A scaling factor to apply to the Y-axis coordinates of the spatial data.</p>
              </div>
            </td>
            <td>
                  <code>1</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>r</code>
            </td>
            <td>
                  <code>float, optional (default: 0.1)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Radius of the pie charts representing cell type fractions at each spatial coordinate.</p>
              </div>
            </td>
            <td>
                  <code>0.1</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>out_dir</code>
            </td>
            <td>
                  <code>string, optional (default: &#39;/plot&#39;)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The directory where the output figure (SVG format) will be saved.</p>
              </div>
            </td>
            <td>
                  <code>&#39;/plot&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>project_name</code>
            </td>
            <td>
                  <code>string, optional (default: &#39;project&#39;)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The prefix to use for the saved SVG file.</p>
              </div>
            </td>
            <td>
                  <code>&#39;project&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>color_list</code>
            </td>
            <td>
                  <code>list, optional (default: None)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A list of colors to use for the cell types. If <code>None</code>, a colormap (specified by <code>color_map</code>) will be used to generate colors for each cell type.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>color_map</code>
            </td>
            <td>
                  <code>string, optional (default: &#39;Spectral&#39;)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The name of the colormap to use if <code>color_list</code> is not provided. It will generate a set of colors for the different cell types.</p>
              </div>
            </td>
            <td>
                  <code>&#39;Spectral&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>rotation_angle</code>
            </td>
            <td>
                  <code>float, optional (default: None)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>An optional angle (in degrees) to rotate the spatial coordinates. If provided, the rotation is applied to the coordinates before plotting.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>figsize</code>
            </td>
            <td>
                  <code>tuple, optional (default: (2, 2))</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The size of the output figure, specified as a tuple of (width, height) in inches.</p>
              </div>
            </td>
            <td>
                  <code>(2, 2)</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>None</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The function does not return any values but saves the generated pie chart plot as an SVG file in the specified <code>out_dir</code>.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>cytobulk\plots\_plot.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">plot_celltype_fraction_pie</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span>
                               <span class="n">scale_facter_x</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
                               <span class="n">scale_factor_y</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
                               <span class="n">r</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
                               <span class="n">out_dir</span><span class="o">=</span><span class="s1">&#39;/plot&#39;</span><span class="p">,</span>
                               <span class="n">project_name</span><span class="o">=</span><span class="s1">&#39;project&#39;</span><span class="p">,</span>
                               <span class="n">color_list</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                               <span class="n">color_map</span><span class="o">=</span><span class="s2">&quot;Spectral&quot;</span><span class="p">,</span>
                               <span class="n">rotation_angle</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                               <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">)):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plot cell type fraction as pie charts at spatial coordinates.</span>

<span class="sd">    This function generates pie charts representing the cell type fraction at each spatial location (from the &#39;spatial&#39; data in `adata`) and plots them on a 2D scatter plot. Each pie chart is positioned based on the spatial coordinates and reflects the relative abundance of each cell type at that spot. The resulting figure is saved as an SVG file.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    adata : anndata.AnnData</span>
<span class="sd">        An :class:`~anndata.AnnData` object containing spatial transcriptomics data.</span>
<span class="sd">        The spatial coordinates should be stored in `adata.obsm[&#39;spatial&#39;]`, and the deconvolution results (cell type fractions) should be stored in `adata.uns[&#39;deconv&#39;]`.</span>

<span class="sd">    scale_facter_x : float, optional (default: 1)</span>
<span class="sd">        A scaling factor to apply to the X-axis coordinates of the spatial data.</span>

<span class="sd">    scale_factor_y : float, optional (default: 1)</span>
<span class="sd">        A scaling factor to apply to the Y-axis coordinates of the spatial data.</span>

<span class="sd">    r : float, optional (default: 0.1)</span>
<span class="sd">        Radius of the pie charts representing cell type fractions at each spatial coordinate.</span>

<span class="sd">    out_dir : string, optional (default: &#39;/plot&#39;)</span>
<span class="sd">        The directory where the output figure (SVG format) will be saved.</span>

<span class="sd">    project_name : string, optional (default: &#39;project&#39;)</span>
<span class="sd">        The prefix to use for the saved SVG file.</span>

<span class="sd">    color_list : list, optional (default: None)</span>
<span class="sd">        A list of colors to use for the cell types. If `None`, a colormap (specified by `color_map`) will be used to generate colors for each cell type.</span>

<span class="sd">    color_map : string, optional (default: &#39;Spectral&#39;)</span>
<span class="sd">        The name of the colormap to use if `color_list` is not provided. It will generate a set of colors for the different cell types.</span>

<span class="sd">    rotation_angle : float, optional (default: None)</span>
<span class="sd">        An optional angle (in degrees) to rotate the spatial coordinates. If provided, the rotation is applied to the coordinates before plotting.</span>

<span class="sd">    figsize : tuple, optional (default: (2, 2))</span>
<span class="sd">        The size of the output figure, specified as a tuple of (width, height) in inches.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    None</span>
<span class="sd">        The function does not return any values but saves the generated pie chart plot as an SVG file in the specified `out_dir`.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">loc_xy</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;spatial&#39;</span><span class="p">]</span>
    <span class="n">loc_xy</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">loc_xy</span><span class="p">,</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;x&#39;</span><span class="p">,</span><span class="s1">&#39;y&#39;</span><span class="p">],</span><span class="n">index</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs_names</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">rotation_angle</span><span class="p">:</span>
        <span class="n">new_loc</span> <span class="o">=</span> <span class="n">loc_xy</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span> <span class="p">:</span> <span class="n">utils</span><span class="o">.</span><span class="n">rotate_matrix</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;x&#39;</span><span class="p">],</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;y&#39;</span><span class="p">],</span><span class="n">rotation_angle</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">loc_xy</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">new_loc</span><span class="p">),</span><span class="n">columns</span><span class="o">=</span><span class="n">loc_xy</span><span class="o">.</span><span class="n">columns</span><span class="p">,</span><span class="n">index</span><span class="o">=</span><span class="n">loc_xy</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
    <span class="n">cell_type_fraction</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;deconv&#39;</span><span class="p">]</span>
    <span class="n">cell_type_fraction</span><span class="p">[</span><span class="s1">&#39;None&#39;</span><span class="p">]</span><span class="o">=</span><span class="mi">0</span>
    <span class="n">cell_type_fraction</span><span class="o">.</span><span class="n">loc</span><span class="p">[(</span><span class="n">cell_type_fraction</span><span class="o">==</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span><span class="s2">&quot;None&quot;</span><span class="p">]</span><span class="o">=</span><span class="mi">1</span>
    <span class="n">cell_type_fraction</span><span class="p">[</span><span class="s1">&#39;x&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">loc_xy</span><span class="p">[</span><span class="s1">&#39;x&#39;</span><span class="p">]</span>
    <span class="n">cell_type_fraction</span><span class="p">[</span><span class="s1">&#39;y&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">loc_xy</span><span class="p">[</span><span class="s1">&#39;y&#39;</span><span class="p">]</span>
    <span class="n">loc</span><span class="o">=</span><span class="n">cell_type_fraction</span>
    <span class="n">loc</span><span class="p">[</span><span class="s1">&#39;max_idx&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">loc</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,:</span><span class="o">-</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">idxmax</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">loc</span><span class="p">[</span><span class="s1">&#39;x&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">loc</span><span class="p">[</span><span class="s1">&#39;x&#39;</span><span class="p">]</span><span class="o">/</span><span class="n">scale_facter_x</span>
    <span class="n">loc</span><span class="p">[</span><span class="s1">&#39;y&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">loc</span><span class="p">[</span><span class="s1">&#39;y&#39;</span><span class="p">]</span><span class="o">/</span><span class="n">scale_factor_y</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">)</span>
    <span class="k">def</span> <span class="nf">_plot_pie</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">ax</span><span class="p">,</span> <span class="n">color_sets</span><span class="p">,</span><span class="n">r</span><span class="p">):</span> 
        <span class="n">fraction</span> <span class="o">=</span> <span class="n">x</span><span class="p">[:</span><span class="o">-</span><span class="mi">3</span><span class="p">]</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">pie</span><span class="p">(</span><span class="n">fraction</span><span class="p">,</span> <span class="n">center</span><span class="o">=</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;x&#39;</span><span class="p">],</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;y&#39;</span><span class="p">]),</span> <span class="n">radius</span><span class="o">=</span><span class="n">r</span><span class="p">,</span><span class="n">colors</span><span class="o">=</span><span class="p">[</span><span class="n">color_sets</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">loc</span><span class="o">.</span><span class="n">columns</span><span class="p">[:</span><span class="o">-</span><span class="mi">3</span><span class="p">]])</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;CytoBulk&#39;</span><span class="p">,</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span><span class="n">x</span><span class="o">=</span><span class="mf">1.2</span><span class="p">,</span><span class="n">y</span><span class="o">=</span><span class="mf">1.7</span><span class="p">)</span>
    <span class="c1"># git min/max values for the axes</span>
    <span class="n">cells</span> <span class="o">=</span> <span class="n">loc</span><span class="o">.</span><span class="n">columns</span><span class="p">[:</span><span class="o">-</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">color_list</span><span class="p">:</span>
        <span class="n">color_dict</span> <span class="o">=</span> <span class="p">{</span><span class="n">item</span><span class="p">:</span> <span class="n">color</span> <span class="k">for</span> <span class="n">item</span><span class="p">,</span> <span class="n">color</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">cells</span><span class="p">,</span> <span class="n">color_list</span><span class="p">)}</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">color_map</span> <span class="o">=</span> <span class="n">cm</span><span class="o">.</span><span class="n">get_cmap</span><span class="p">(</span><span class="n">color_map</span><span class="p">)</span>
        <span class="n">list_length</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">cells</span><span class="p">)</span>
        <span class="n">colors_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">to_hex</span><span class="p">(</span><span class="n">color_map</span><span class="p">(</span><span class="n">i</span> <span class="o">/</span> <span class="p">(</span><span class="n">list_length</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)))</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">list_length</span><span class="p">)]</span>
        <span class="n">color_dict</span> <span class="o">=</span> <span class="p">{</span><span class="n">item</span><span class="p">:</span> <span class="n">color</span> <span class="k">for</span> <span class="n">item</span><span class="p">,</span> <span class="n">color</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">cells</span><span class="p">,</span> <span class="n">colors_list</span><span class="p">)}</span>

    <span class="n">loc</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span> <span class="p">:</span> <span class="n">_plot_pie</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">ax</span><span class="p">,</span><span class="n">color_dict</span><span class="p">,</span><span class="n">r</span><span class="o">=</span><span class="mf">0.1</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="n">handles</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">l</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">cells</span><span class="p">):</span>
        <span class="n">handles</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">matplotlib</span><span class="o">.</span><span class="n">patches</span><span class="o">.</span><span class="n">Patch</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="n">colors_list</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">label</span><span class="o">=</span><span class="n">cells</span><span class="p">[</span><span class="n">i</span><span class="p">]))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">handles</span><span class="p">,</span><span class="n">cells</span><span class="p">,</span> <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mf">0.1</span><span class="p">,</span><span class="mf">0.1</span><span class="p">),</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span><span class="n">markerscale</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span><span class="n">ncol</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span><span class="n">loc</span><span class="o">=</span><span class="s2">&quot;upper left&quot;</span><span class="p">)</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">out_dir</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">project_name</span><span class="si">}</span><span class="s2">_cell_fraction_pie.svg&quot;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;svg&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">1200</span><span class="p">,</span><span class="n">bbox_inches</span> <span class="o">=</span> <span class="s1">&#39;tight&#39;</span><span class="p">,</span><span class="n">transparent</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div><hr />
<p><div class="highlight"><pre><span></span><code><span class="n">plot_celltype_fraction_heatmap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span>
                                   <span class="n">label</span><span class="p">,</span>
                                    <span class="n">r</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
                                    <span class="n">out_dir</span><span class="o">=</span><span class="s1">&#39;/plot&#39;</span><span class="p">,</span>
                                    <span class="n">project_name</span><span class="o">=</span><span class="s1">&#39;project&#39;</span><span class="p">,</span>
                                    <span class="n">color_map</span><span class="o">=</span><span class="s1">&#39;crest&#39;</span><span class="p">,</span>
                                    <span class="n">rotation_angle</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                                    <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mf">2.7</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span>
</code></pre></div>
</p>


<div class="doc doc-object doc-function">



<a id="cytobulk.plots._plot.plot_celltype_fraction_heatmap"></a>
    <div class="doc doc-contents first">

        <p>Plot a heatmap of cell type fractions at spatial coordinates.</p>
<p>This function visualizes the fraction of a specified cell type as a heatmap on a 2D scatter plot, using spatial transcriptomics data. The color intensity of each point corresponds to the cell type fraction at that spatial location. The resulting figure is saved as an SVG file.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>adata</code>
            </td>
            <td>
                  <code><span title="anndata.AnnData">AnnData</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>An :class:<code>~anndata.AnnData</code> object containing spatial transcriptomics data.
The spatial coordinates should be stored in <code>adata.obsm['spatial']</code>, and the deconvolution results (cell type fractions) should be stored in <code>adata.uns['deconv']</code>.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>label</code>
            </td>
            <td>
                  <code>string</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The name of the cell type to plot. This should correspond to one of the columns in <code>adata.uns['deconv']</code>, which represents the fractions of different cell types.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>r</code>
            </td>
            <td>
                  <code>float, optional (default: 0.1)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Radius of the points representing the cell type fractions at each spatial coordinate.</p>
              </div>
            </td>
            <td>
                  <code>0.1</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>out_dir</code>
            </td>
            <td>
                  <code>string, optional (default: &#39;/plot&#39;)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The directory where the output figure (SVG format) will be saved.</p>
              </div>
            </td>
            <td>
                  <code>&#39;/plot&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>project_name</code>
            </td>
            <td>
                  <code>string, optional (default: &#39;project&#39;)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The prefix to use for the saved SVG file.</p>
              </div>
            </td>
            <td>
                  <code>&#39;project&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>color_map</code>
            </td>
            <td>
                  <code>string, optional (default: &#39;crest&#39;)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The name of the colormap to use for the heatmap. This can be any colormap recognized by <code>seaborn</code> or <code>matplotlib</code>.</p>
              </div>
            </td>
            <td>
                  <code>&#39;crest&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>rotation_angle</code>
            </td>
            <td>
                  <code>float, optional (default: None)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>An optional angle (in degrees) to rotate the spatial coordinates. If provided, the rotation is applied to the coordinates before plotting.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>figsize</code>
            </td>
            <td>
                  <code>tuple, optional (default: (2.7, 2))</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The size of the output figure, specified as a tuple of (width, height) in inches.</p>
              </div>
            </td>
            <td>
                  <code>(2.7, 2)</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>None</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The function does not return any values but saves the generated heatmap plot as an SVG file in the specified <code>out_dir</code>.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>cytobulk\plots\_plot.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">plot_celltype_fraction_heatmap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span>
                                   <span class="n">label</span><span class="p">,</span>
                                    <span class="n">r</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
                                    <span class="n">out_dir</span><span class="o">=</span><span class="s1">&#39;/plot&#39;</span><span class="p">,</span>
                                    <span class="n">project_name</span><span class="o">=</span><span class="s1">&#39;project&#39;</span><span class="p">,</span>
                                    <span class="n">color_map</span><span class="o">=</span><span class="s1">&#39;crest&#39;</span><span class="p">,</span>
                                    <span class="n">rotation_angle</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                                    <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mf">2.7</span><span class="p">,</span> <span class="mi">2</span><span class="p">)):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plot a heatmap of cell type fractions at spatial coordinates.</span>

<span class="sd">    This function visualizes the fraction of a specified cell type as a heatmap on a 2D scatter plot, using spatial transcriptomics data. The color intensity of each point corresponds to the cell type fraction at that spatial location. The resulting figure is saved as an SVG file.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    adata : anndata.AnnData</span>
<span class="sd">        An :class:`~anndata.AnnData` object containing spatial transcriptomics data.</span>
<span class="sd">        The spatial coordinates should be stored in `adata.obsm[&#39;spatial&#39;]`, and the deconvolution results (cell type fractions) should be stored in `adata.uns[&#39;deconv&#39;]`.</span>

<span class="sd">    label : string</span>
<span class="sd">        The name of the cell type to plot. This should correspond to one of the columns in `adata.uns[&#39;deconv&#39;]`, which represents the fractions of different cell types.</span>

<span class="sd">    r : float, optional (default: 0.1)</span>
<span class="sd">        Radius of the points representing the cell type fractions at each spatial coordinate.</span>

<span class="sd">    out_dir : string, optional (default: &#39;/plot&#39;)</span>
<span class="sd">        The directory where the output figure (SVG format) will be saved.</span>

<span class="sd">    project_name : string, optional (default: &#39;project&#39;)</span>
<span class="sd">        The prefix to use for the saved SVG file.</span>

<span class="sd">    color_map : string, optional (default: &#39;crest&#39;)</span>
<span class="sd">        The name of the colormap to use for the heatmap. This can be any colormap recognized by `seaborn` or `matplotlib`.</span>

<span class="sd">    rotation_angle : float, optional (default: None)</span>
<span class="sd">        An optional angle (in degrees) to rotate the spatial coordinates. If provided, the rotation is applied to the coordinates before plotting.</span>

<span class="sd">    figsize : tuple, optional (default: (2.7, 2))</span>
<span class="sd">        The size of the output figure, specified as a tuple of (width, height) in inches.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    None</span>
<span class="sd">        The function does not return any values but saves the generated heatmap plot as an SVG file in the specified `out_dir`.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="kn">from</span> <span class="nn">sklearn.preprocessing</span> <span class="kn">import</span> <span class="n">minmax_scale</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">([])</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">yticks</span><span class="p">([])</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
    <span class="n">loc_xy</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;spatial&#39;</span><span class="p">]</span>
    <span class="n">loc_xy</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">loc_xy</span><span class="p">,</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;x&#39;</span><span class="p">,</span><span class="s1">&#39;y&#39;</span><span class="p">],</span><span class="n">index</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs_names</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">rotation_angle</span><span class="p">:</span>
        <span class="n">new_loc</span> <span class="o">=</span> <span class="n">loc_xy</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span> <span class="p">:</span> <span class="n">utils</span><span class="o">.</span><span class="n">rotate_matrix</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;x&#39;</span><span class="p">],</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;y&#39;</span><span class="p">],</span><span class="n">rotation_angle</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">loc_xy</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">new_loc</span><span class="p">),</span><span class="n">columns</span><span class="o">=</span><span class="n">loc_xy</span><span class="o">.</span><span class="n">columns</span><span class="p">,</span><span class="n">index</span><span class="o">=</span><span class="n">loc_xy</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
    <span class="n">cell_type_fraction</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;deconv&#39;</span><span class="p">]</span>
    <span class="n">cell_type_fraction</span><span class="p">[</span><span class="s1">&#39;x&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">loc_xy</span><span class="p">[</span><span class="s1">&#39;x&#39;</span><span class="p">]</span>
    <span class="n">cell_type_fraction</span><span class="p">[</span><span class="s1">&#39;y&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">loc_xy</span><span class="p">[</span><span class="s1">&#39;y&#39;</span><span class="p">]</span>
    <span class="n">cell_type_fraction</span><span class="p">[</span><span class="n">label</span><span class="p">]</span> <span class="o">=</span> <span class="n">minmax_scale</span><span class="p">(</span><span class="n">cell_type_fraction</span><span class="p">[</span><span class="n">label</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">cell_type_fraction</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s2">&quot;y&quot;</span><span class="p">,</span><span class="n">palette</span><span class="o">=</span><span class="n">color_map</span><span class="p">,</span><span class="n">hue</span><span class="o">=</span><span class="n">label</span><span class="p">,</span><span class="n">s</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span><span class="n">legend</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">edgecolor</span><span class="o">=</span><span class="s2">&quot;None&quot;</span><span class="p">,</span><span class="n">markers</span><span class="o">=</span><span class="s2">&quot;pentagon&quot;</span><span class="p">)</span>
    <span class="n">norm</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">Normalize</span><span class="p">(</span><span class="n">cell_type_fraction</span><span class="p">[</span><span class="n">label</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">(),</span> <span class="n">cell_type_fraction</span><span class="p">[</span><span class="n">label</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">())</span>
    <span class="n">cmap</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">(</span><span class="n">color_map</span><span class="p">,</span> <span class="n">as_cmap</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">sm</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">cm</span><span class="o">.</span><span class="n">ScalarMappable</span><span class="p">(</span><span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span><span class="p">,</span> <span class="n">norm</span><span class="o">=</span><span class="n">norm</span><span class="p">)</span>
    <span class="n">sm</span><span class="o">.</span><span class="n">set_array</span><span class="p">([])</span>
    <span class="n">cax</span> <span class="o">=</span> <span class="n">fig</span><span class="o">.</span><span class="n">add_axes</span><span class="p">([</span><span class="n">ax</span><span class="o">.</span><span class="n">get_position</span><span class="p">()</span><span class="o">.</span><span class="n">x1</span><span class="o">+</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">ax</span><span class="o">.</span><span class="n">get_position</span><span class="p">()</span><span class="o">.</span><span class="n">y0</span><span class="p">,</span> <span class="mf">0.03</span><span class="p">,</span> <span class="n">ax</span><span class="o">.</span><span class="n">get_position</span><span class="p">()</span><span class="o">.</span><span class="n">height</span><span class="p">])</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">label</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span><span class="n">x</span><span class="o">=</span><span class="n">ax</span><span class="o">.</span><span class="n">get_position</span><span class="p">()</span><span class="o">.</span><span class="n">x1</span><span class="o">/</span><span class="mi">2</span><span class="p">,</span><span class="n">y</span><span class="o">=</span><span class="n">ax</span><span class="o">.</span><span class="n">get_position</span><span class="p">()</span><span class="o">.</span><span class="n">y1</span><span class="o">+</span><span class="mf">0.1</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">figure</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">sm</span><span class="p">,</span> <span class="n">cax</span><span class="o">=</span><span class="n">cax</span><span class="p">)</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">out_dir</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">project_name</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">label</span><span class="si">}</span><span class="s2">.svg&quot;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;svg&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">1200</span><span class="p">,</span><span class="n">bbox_inches</span> <span class="o">=</span> <span class="s1">&#39;tight&#39;</span><span class="p">,</span><span class="n">transparent</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div><hr />
<p><div class="highlight"><pre><span></span><code><span class="n">plot_paired_violin</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span>
                       <span class="n">label</span><span class="p">,</span>
                       <span class="n">gene</span><span class="p">,</span>
                       <span class="n">stats_method</span><span class="o">=</span><span class="s1">&#39;spearmanr&#39;</span><span class="p">,</span>
                       <span class="n">out_dir</span><span class="o">=</span><span class="s1">&#39;/plot&#39;</span><span class="p">,</span>
                       <span class="n">color_list</span><span class="o">=</span><span class="n">Const</span><span class="o">.</span><span class="n">DIFF_COLOR_N</span><span class="p">,</span>
                       <span class="n">project_name</span><span class="o">=</span><span class="s2">&quot;test&quot;</span><span class="p">,</span>
                       <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span>
                       <span class="n">ylim</span><span class="o">=</span><span class="p">[</span><span class="o">-</span><span class="mf">0.1</span><span class="p">,</span> <span class="mf">1.2</span><span class="p">])</span>
</code></pre></div>
</p>


<div class="doc doc-object doc-function">



<a id="cytobulk.plots._plot.plot_paired_violin"></a>
    <div class="doc doc-contents first">

        <p>Plot paired violin plots for gene expression and predicted cell type fractions.</p>
<p>This function generates paired violin plots comparing the expression of a specific gene with the predicted cell type fraction (from deconvolution data) for the same set of observations. It also computes a statistical correlation (Spearman or Pearson) between the gene expression and the predicted cell type fractions and displays the correlation coefficient and significance on the plot.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>adata</code>
            </td>
            <td>
                  <code><span title="anndata.AnnData">AnnData</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>An :class:<code>~anndata.AnnData</code> object containing gene expression data and the output of a cell type deconvolution.
- Gene expression data should be stored in <code>adata.X</code> with cell/barcode names in <code>adata.obs_names</code> and gene names in <code>adata.var_names</code>.
- Predicted cell type fractions (from deconvolution) should be stored in <code>adata.uns['deconv']</code>.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>label</code>
            </td>
            <td>
                  <code>string</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The name of the cell type (as stored in <code>adata.uns['deconv']</code>) to be compared against the gene expression.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>gene</code>
            </td>
            <td>
                  <code>string</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The name of the gene (as stored in <code>adata.var_names</code>) whose expression levels will be compared to the predicted cell type fractions.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>stats_method</code>
            </td>
            <td>
                  <code>string, optional (default: &#39;spearmanr&#39;)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The statistical method used to compute the correlation between gene expression and predicted cell type fractions.
- <code>'spearmanr'</code>: Spearmans rank correlation.
- <code>'pearsonr'</code>: Pearsons correlation.</p>
              </div>
            </td>
            <td>
                  <code>&#39;spearmanr&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>out_dir</code>
            </td>
            <td>
                  <code>string, optional (default: &#39;/plot&#39;)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The directory where the output violin plot (SVG format) will be saved.</p>
              </div>
            </td>
            <td>
                  <code>&#39;/plot&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>color_list</code>
            </td>
            <td>
                  <code>list, optional (default: Const.DIFF_COLOR_N)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A list of colors to use for the violin plots. If not provided, a default color palette will be used.</p>
              </div>
            </td>
            <td>
                  <code><span title="cytobulk.plots._plot.Const.DIFF_COLOR_N">DIFF_COLOR_N</span></code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>project_name</code>
            </td>
            <td>
                  <code>string, optional (default: &#39;test&#39;)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The prefix to use for the saved SVG file.</p>
              </div>
            </td>
            <td>
                  <code>&#39;test&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>figsize</code>
            </td>
            <td>
                  <code>tuple, optional (default: (6, 4))</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The size of the output figure, specified as a tuple of (width, height) in inches.</p>
              </div>
            </td>
            <td>
                  <code>(6, 4)</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>ylim</code>
            </td>
            <td>
                  <code>list, optional (default: [-0.1, 1.2])</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The limits for the Y-axis of the violin plot. This should be a list of two values [min, max].</p>
              </div>
            </td>
            <td>
                  <code>[-0.1, 1.2]</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>None</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The function does not return any values but saves the generated violin plot as an SVG file in the specified <code>out_dir</code>.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>cytobulk\plots\_plot.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">plot_paired_violin</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span>
                       <span class="n">label</span><span class="p">,</span>
                       <span class="n">gene</span><span class="p">,</span>
                       <span class="n">stats_method</span><span class="o">=</span><span class="s1">&#39;spearmanr&#39;</span><span class="p">,</span>
                       <span class="n">out_dir</span><span class="o">=</span><span class="s1">&#39;/plot&#39;</span><span class="p">,</span>
                       <span class="n">color_list</span><span class="o">=</span><span class="n">Const</span><span class="o">.</span><span class="n">DIFF_COLOR_N</span><span class="p">,</span>
                       <span class="n">project_name</span><span class="o">=</span><span class="s2">&quot;test&quot;</span><span class="p">,</span>
                       <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span>
                       <span class="n">ylim</span><span class="o">=</span><span class="p">[</span><span class="o">-</span><span class="mf">0.1</span><span class="p">,</span> <span class="mf">1.2</span><span class="p">]):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plot paired violin plots for gene expression and predicted cell type fractions.</span>

<span class="sd">    This function generates paired violin plots comparing the expression of a specific gene with the predicted cell type fraction (from deconvolution data) for the same set of observations. It also computes a statistical correlation (Spearman or Pearson) between the gene expression and the predicted cell type fractions and displays the correlation coefficient and significance on the plot.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    adata : anndata.AnnData</span>
<span class="sd">        An :class:`~anndata.AnnData` object containing gene expression data and the output of a cell type deconvolution.</span>
<span class="sd">        - Gene expression data should be stored in `adata.X` with cell/barcode names in `adata.obs_names` and gene names in `adata.var_names`.</span>
<span class="sd">        - Predicted cell type fractions (from deconvolution) should be stored in `adata.uns[&#39;deconv&#39;]`.</span>

<span class="sd">    label : string</span>
<span class="sd">        The name of the cell type (as stored in `adata.uns[&#39;deconv&#39;]`) to be compared against the gene expression.</span>

<span class="sd">    gene : string</span>
<span class="sd">        The name of the gene (as stored in `adata.var_names`) whose expression levels will be compared to the predicted cell type fractions.</span>

<span class="sd">    stats_method : string, optional (default: &#39;spearmanr&#39;)</span>
<span class="sd">        The statistical method used to compute the correlation between gene expression and predicted cell type fractions.</span>
<span class="sd">        - `&#39;spearmanr&#39;`: Spearmans rank correlation.</span>
<span class="sd">        - `&#39;pearsonr&#39;`: Pearsons correlation.</span>

<span class="sd">    out_dir : string, optional (default: &#39;/plot&#39;)</span>
<span class="sd">        The directory where the output violin plot (SVG format) will be saved.</span>

<span class="sd">    color_list : list, optional (default: Const.DIFF_COLOR_N)</span>
<span class="sd">        A list of colors to use for the violin plots. If not provided, a default color palette will be used.</span>

<span class="sd">    project_name : string, optional (default: &#39;test&#39;)</span>
<span class="sd">        The prefix to use for the saved SVG file.</span>

<span class="sd">    figsize : tuple, optional (default: (6, 4))</span>
<span class="sd">        The size of the output figure, specified as a tuple of (width, height) in inches.</span>

<span class="sd">    ylim : list, optional (default: [-0.1, 1.2])</span>
<span class="sd">        The limits for the Y-axis of the violin plot. This should be a list of two values [min, max].</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    None</span>
<span class="sd">        The function does not return any values but saves the generated violin plot as an SVG file in the specified `out_dir`.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="kn">from</span> <span class="nn">scipy</span> <span class="kn">import</span> <span class="n">stats</span>
    <span class="kn">from</span> <span class="nn">sklearn.preprocessing</span> <span class="kn">import</span> <span class="n">minmax_scale</span>
    <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">)</span>
    <span class="n">color_panel</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">set_palette</span><span class="p">(</span><span class="n">color_list</span><span class="p">)</span>
    <span class="n">name_list</span><span class="o">=</span><span class="p">[]</span>
    <span class="n">value_list</span><span class="o">=</span><span class="p">[]</span>
    <span class="n">exp_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">,</span><span class="n">index</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs_names</span><span class="p">,</span><span class="n">columns</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">)</span>
    <span class="n">predicted</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;deconv&#39;</span><span class="p">]</span>
    <span class="n">common_label</span><span class="o">=</span><span class="n">exp_data</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">predicted</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
    <span class="n">predicted</span><span class="o">=</span><span class="n">predicted</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">common_label</span><span class="p">,:]</span>
    <span class="n">exp_data</span><span class="o">=</span><span class="n">exp_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">common_label</span><span class="p">,:]</span>
    <span class="n">exp_data</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span> <span class="o">=</span> <span class="n">minmax_scale</span><span class="p">((</span><span class="n">exp_data</span><span class="p">[</span><span class="n">gene</span><span class="p">]),</span> <span class="n">feature_range</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
    <span class="n">predicted</span><span class="p">[</span><span class="n">label</span><span class="p">]</span> <span class="o">=</span> <span class="n">minmax_scale</span><span class="p">(</span><span class="n">predicted</span><span class="p">[</span><span class="n">label</span><span class="p">],</span> <span class="n">feature_range</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
    <span class="k">if</span> <span class="n">stats_method</span><span class="o">==</span><span class="s1">&#39;spearmanr&#39;</span><span class="p">:</span>
        <span class="n">stat</span><span class="p">,</span><span class="n">p_value</span> <span class="o">=</span> <span class="n">stats</span><span class="o">.</span><span class="n">spearmanr</span><span class="p">(</span><span class="n">exp_data</span><span class="p">[</span><span class="n">gene</span><span class="p">],</span><span class="n">predicted</span><span class="p">[</span><span class="n">label</span><span class="p">])</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">stat</span><span class="p">,</span><span class="n">p_value</span> <span class="o">=</span> <span class="n">stats</span><span class="o">.</span><span class="n">pearsonr</span><span class="p">(</span><span class="n">exp_data</span><span class="p">[</span><span class="n">gene</span><span class="p">],</span><span class="n">predicted</span><span class="p">[</span><span class="n">label</span><span class="p">])</span>
    <span class="n">p_value</span><span class="o">=</span><span class="n">utils</span><span class="o">.</span><span class="n">convert_pvalue_to_asterisks</span><span class="p">(</span><span class="n">p_value</span><span class="p">)</span>
    <span class="n">value_list</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">exp_data</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
    <span class="n">name_list</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="n">gene</span><span class="p">]</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">exp_data</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">))</span>
    <span class="n">value_list</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">predicted</span><span class="p">[</span><span class="n">label</span><span class="p">])</span>
    <span class="n">name_list</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="n">label</span><span class="p">]</span><span class="o">*</span><span class="nb">len</span><span class="p">(</span><span class="n">predicted</span><span class="p">[</span><span class="n">label</span><span class="p">]))</span>
    <span class="n">input_data</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_dict</span><span class="p">({</span><span class="s1">&#39;source&#39;</span><span class="p">:</span><span class="n">name_list</span><span class="p">,</span><span class="s1">&#39;value&#39;</span><span class="p">:</span><span class="n">value_list</span><span class="p">})</span>
    <span class="n">ax</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">violinplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s2">&quot;source&quot;</span><span class="p">,</span><span class="n">y</span><span class="o">=</span><span class="s2">&quot;value&quot;</span><span class="p">,</span><span class="n">data</span><span class="o">=</span><span class="n">input_data</span><span class="p">,</span><span class="n">hue</span><span class="o">=</span><span class="s1">&#39;source&#39;</span><span class="p">,</span><span class="n">palette</span><span class="o">=</span><span class="n">color_panel</span><span class="p">)</span>
    <span class="n">x1</span><span class="p">,</span> <span class="n">x2</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span>
    <span class="n">y</span><span class="p">,</span><span class="n">h</span> <span class="o">=</span> <span class="n">predicted</span><span class="p">[</span><span class="n">label</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">()</span><span class="o">+</span><span class="mf">.1</span><span class="p">,</span><span class="mf">.1</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">x1</span><span class="p">,</span> <span class="n">x1</span><span class="p">,</span> <span class="n">x2</span><span class="p">,</span> <span class="n">x2</span><span class="p">],</span> <span class="p">[</span><span class="n">y</span><span class="p">,</span> <span class="n">y</span><span class="o">+</span><span class="n">h</span><span class="p">,</span> <span class="n">y</span><span class="o">+</span><span class="n">h</span><span class="p">,</span> <span class="n">y</span><span class="p">],</span> <span class="n">lw</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="s2">&quot;k&quot;</span><span class="p">)</span> 
    <span class="n">ax</span><span class="o">.</span><span class="n">text</span><span class="p">((</span><span class="n">x1</span><span class="o">+</span><span class="n">x2</span><span class="p">)</span><span class="o">*</span><span class="mf">.5</span><span class="p">,</span> <span class="n">y</span><span class="o">+</span><span class="n">h</span><span class="p">,</span> <span class="n">stats_method</span><span class="o">+</span><span class="s2">&quot;: &quot;</span><span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%.2f</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">stat</span><span class="p">)</span> <span class="o">+</span><span class="s2">&quot; &quot;</span><span class="o">+</span>  <span class="n">p_value</span><span class="p">,</span> <span class="n">ha</span><span class="o">=</span><span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">va</span><span class="o">=</span><span class="s1">&#39;bottom&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;k&quot;</span><span class="p">,</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">14</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">([</span><span class="o">-</span><span class="mf">0.3</span><span class="p">,</span> <span class="mf">1.5</span><span class="p">])</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">which</span><span class="o">=</span><span class="s1">&#39;major&#39;</span><span class="p">,</span><span class="n">direction</span><span class="o">=</span><span class="s1">&#39;in&#39;</span><span class="p">,</span><span class="n">length</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span><span class="n">width</span><span class="o">=</span><span class="mf">1.</span><span class="p">,</span><span class="n">labelsize</span><span class="o">=</span><span class="mi">14</span><span class="p">,</span><span class="n">bottom</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">out_dir</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">project_name</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">label</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">_violin.svg&quot;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;svg&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">1200</span><span class="p">,</span><span class="n">bbox_inches</span> <span class="o">=</span> <span class="s1">&#39;tight&#39;</span><span class="p">,</span><span class="n">transparent</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div><hr />
<p><div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">plot_reconstruction</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span>
                        <span class="n">out_dir</span><span class="p">,</span>
                        <span class="n">project_name</span><span class="o">=</span><span class="s2">&quot;test&quot;</span><span class="p">,</span>
                        <span class="n">rotation_angle</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                        <span class="n">spot_size</span><span class="o">=</span><span class="mf">0.5</span><span class="p">):</span>
</code></pre></div>
</p>


<div class="doc doc-object doc-function">



<a id="cytobulk.plots._plot.plot_reconstruction"></a>
    <div class="doc doc-contents first">

        <p>Plot reconstructed spatial transcriptomics data with correlation analysis.</p>
<p>This function computes the Pearson correlation between the original and reconstructed spatial transcriptomics data for each spot. It then visualizes the spatial distribution of correlation coefficients using a scatter plot.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>adata</code>
            </td>
            <td>
                  <code><span title="anndata.AnnData">AnnData</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>An :class:<code>~anndata.AnnData</code> object containing the spatial transcriptomics data.
- Original data should be in <code>adata.layers['original_st']</code>.
- Reconstructed data should be in <code>adata.X</code>.
- Spatial coordinates should be in <code>adata.obsm['spatial']</code>.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>out_dir</code>
            </td>
            <td>
                  <code>string</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The directory where the output plot (SVG format) will be saved.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>project_name</code>
            </td>
            <td>
                  <code>string, optional (default: &#39;test&#39;)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The prefix to use for the saved SVG file.</p>
              </div>
            </td>
            <td>
                  <code>&#39;test&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>rotation_angle</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The angle to rotate the spatial coordinates, if any.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>spot_size</code>
            </td>
            <td>
                  <code>float, optional (default: 0.5)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The size of the spots in the scatter plot.</p>
              </div>
            </td>
            <td>
                  <code>0.5</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>None</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The function does not return any values but saves the generated scatter plot as an SVG file in the specified <code>out_dir</code>.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>cytobulk\plots\_plot.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">plot_reconstruction</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span>
                        <span class="n">out_dir</span><span class="p">,</span>
                        <span class="n">project_name</span><span class="o">=</span><span class="s2">&quot;test&quot;</span><span class="p">,</span>
                        <span class="n">rotation_angle</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                        <span class="n">spot_size</span><span class="o">=</span><span class="mf">0.5</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plot reconstructed spatial transcriptomics data with correlation analysis.</span>

<span class="sd">    This function computes the Pearson correlation between the original and reconstructed spatial transcriptomics data for each spot. It then visualizes the spatial distribution of correlation coefficients using a scatter plot.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    adata : anndata.AnnData</span>
<span class="sd">        An :class:`~anndata.AnnData` object containing the spatial transcriptomics data.</span>
<span class="sd">        - Original data should be in `adata.layers[&#39;original_st&#39;]`.</span>
<span class="sd">        - Reconstructed data should be in `adata.X`.</span>
<span class="sd">        - Spatial coordinates should be in `adata.obsm[&#39;spatial&#39;]`.</span>

<span class="sd">    out_dir : string</span>
<span class="sd">        The directory where the output plot (SVG format) will be saved.</span>

<span class="sd">    project_name : string, optional (default: &#39;test&#39;)</span>
<span class="sd">        The prefix to use for the saved SVG file.</span>

<span class="sd">    rotation_angle : float, optional</span>
<span class="sd">        The angle to rotate the spatial coordinates, if any.</span>

<span class="sd">    spot_size : float, optional (default: 0.5)</span>
<span class="sd">        The size of the spots in the scatter plot.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    None</span>
<span class="sd">        The function does not return any values but saves the generated scatter plot as an SVG file in the specified `out_dir`.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="kn">from</span> <span class="nn">scipy.stats</span> <span class="kn">import</span> <span class="n">pearsonr</span>
    <span class="n">adata</span><span class="o">.</span><span class="n">var_names_make_unique</span><span class="p">()</span>
    <span class="c1"># Calculate Pearson correlation and p-value for each sample (row)</span>
    <span class="n">loc_xy</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;spatial&#39;</span><span class="p">]</span>
    <span class="n">loc_xy</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">loc_xy</span><span class="p">,</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;x&#39;</span><span class="p">,</span><span class="s1">&#39;y&#39;</span><span class="p">],</span><span class="n">index</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs_names</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">rotation_angle</span><span class="p">:</span>
        <span class="n">new_loc</span> <span class="o">=</span> <span class="n">loc_xy</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span> <span class="p">:</span> <span class="n">utils</span><span class="o">.</span><span class="n">rotate_matrix</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;x&#39;</span><span class="p">],</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;y&#39;</span><span class="p">],</span><span class="n">rotation_angle</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">loc_xy</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">new_loc</span><span class="p">),</span><span class="n">columns</span><span class="o">=</span><span class="n">loc_xy</span><span class="o">.</span><span class="n">columns</span><span class="p">,</span><span class="n">index</span><span class="o">=</span><span class="n">loc_xy</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>

    <span class="n">results</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>  <span class="c1"># Loop through each sample</span>
        <span class="n">r</span><span class="p">,</span> <span class="n">p</span> <span class="o">=</span> <span class="n">pearsonr</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s1">&#39;original_st&#39;</span><span class="p">][</span><span class="n">i</span><span class="p">],</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
        <span class="n">results</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">i</span><span class="p">,</span> <span class="n">r</span><span class="p">,</span> <span class="n">p</span><span class="p">))</span>  <span class="c1"># Store index, Pearson r, and p-value</span>

    <span class="c1"># Convert results to a DataFrame</span>
    <span class="n">results_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">results</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;spot&#39;</span><span class="p">,</span> <span class="s1">&#39;Pearson R&#39;</span><span class="p">,</span> <span class="s1">&#39;pvalue&#39;</span><span class="p">])</span>
    <span class="n">results_df</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;spot&#39;</span><span class="p">,</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;spatial&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">loc_xy</span><span class="o">.</span><span class="n">values</span>
    <span class="n">adata</span><span class="o">.</span><span class="n">obs</span> <span class="o">=</span> <span class="n">results_df</span>
    <span class="n">average_corr</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">results_df</span><span class="p">[</span><span class="s1">&#39;Pearson R&#39;</span><span class="p">]),</span><span class="mi">3</span><span class="p">)</span>

    <span class="n">fig</span><span class="o">=</span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spatial</span><span class="p">(</span>
        <span class="n">adata</span><span class="p">,</span>
        <span class="n">color</span><span class="o">=</span><span class="s1">&#39;Pearson R&#39;</span><span class="p">,</span>
        <span class="n">img_key</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">alpha</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span>
        <span class="n">color_map</span><span class="o">=</span><span class="s2">&quot;mako_r&quot;</span><span class="p">,</span>
        <span class="n">size</span><span class="o">=</span><span class="mf">1.5</span><span class="p">,</span>
        <span class="n">title</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;Reconstructed ST</span><span class="se">\n</span><span class="s1">mean Pearson correlation = </span><span class="si">{</span><span class="n">average_corr</span><span class="si">}</span><span class="se">\n</span><span class="s1">gene number = </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">)</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span>
        <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">spot_size</span><span class="o">=</span><span class="n">spot_size</span><span class="p">,</span>
        <span class="n">outline_width</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
        <span class="n">return_fig</span><span class="o">=</span><span class="kc">True</span>
    <span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">out_dir</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">project_name</span><span class="si">}</span><span class="s2">_reconstructed_correlation.svg&quot;</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;svg&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">1200</span><span class="p">,</span><span class="n">bbox_inches</span> <span class="o">=</span> <span class="s1">&#39;tight&#39;</span><span class="p">,</span><span class="n">transparent</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div><hr />
<p><div class="highlight"><pre><span></span><code><span class="n">df</span> <span class="n">plot_gene_similarity</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> 
                        <span class="n">marker_df</span><span class="p">,</span> 
                        <span class="n">custom_palette</span><span class="o">=</span><span class="s2">&quot;Spectral&quot;</span><span class="p">)</span>
</code></pre></div>
</p>


<div class="doc doc-object doc-function">



<a id="cytobulk.plots._plot.plot_gene_similarity"></a>
    <div class="doc doc-contents first">

        <p>Compute and visualize cosine similarity for marker genes across cell types.</p>
<p>This function calculates cosine similarity for marker genes between the gene expression matrix 
and a mapping layer in an AnnData object. It generates a swarm plot to visualize the results 
grouped by cell type.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>adata</code>
            </td>
            <td>
                  <code><span title="anndata.AnnData">AnnData</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>An :class:<code>~anndata.AnnData</code> object containing gene expression data in <code>.X</code> and a mapping layer in <code>.layers["mapping_ori"]</code>.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>marker_df</code>
            </td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>DataFrame containing marker gene information. It must include the following columns:
- 'gene_symbol': Gene symbol names.
- 'cell_type': Cell type associated with the marker gene.
- 'pvalue': P-value indicating marker significance.
- 'score': Score representing the marker strength.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>custom_palette</code>
            </td>
            <td>
                  <code>str or list, optional (default: &#34;Spectral&#34;)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Custom color palette for plotting. Can be a string representing a seaborn palette or a list of colors.</p>
              </div>
            </td>
            <td>
                  <code>&#39;Spectral&#39;</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="matplotlib.pyplot.Figure">Figure</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A swarm plot showing cosine similarities for marker genes grouped by cell type.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>cytobulk\plots\_plot.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">plot_gene_similarity</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">marker_df</span><span class="p">,</span> <span class="n">custom_palette</span><span class="o">=</span><span class="s2">&quot;Spectral&quot;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Compute and visualize cosine similarity for marker genes across cell types.</span>

<span class="sd">    This function calculates cosine similarity for marker genes between the gene expression matrix </span>
<span class="sd">    and a mapping layer in an AnnData object. It generates a swarm plot to visualize the results </span>
<span class="sd">    grouped by cell type.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    adata : anndata.AnnData</span>
<span class="sd">        An :class:`~anndata.AnnData` object containing gene expression data in `.X` and a mapping layer in `.layers[&quot;mapping_ori&quot;]`.</span>

<span class="sd">    marker_df : pandas.DataFrame</span>
<span class="sd">        DataFrame containing marker gene information. It must include the following columns:</span>
<span class="sd">        - &#39;gene_symbol&#39;: Gene symbol names.</span>
<span class="sd">        - &#39;cell_type&#39;: Cell type associated with the marker gene.</span>
<span class="sd">        - &#39;pvalue&#39;: P-value indicating marker significance.</span>
<span class="sd">        - &#39;score&#39;: Score representing the marker strength.</span>

<span class="sd">    custom_palette : str or list, optional (default: &quot;Spectral&quot;)</span>
<span class="sd">        Custom color palette for plotting. Can be a string representing a seaborn palette or a list of colors.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    matplotlib.pyplot.Figure</span>
<span class="sd">        A swarm plot showing cosine similarities for marker genes grouped by cell type.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Access the gene expression matrix and mapping layer</span>
    <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
    <span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
    <span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
    <span class="kn">from</span> <span class="nn">scipy.spatial.distance</span> <span class="kn">import</span> <span class="n">cosine</span>
    <span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
    <span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
    <span class="n">mapping_ori</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;mapping_ori&quot;</span><span class="p">]</span>  <span class="c1"># Original mapping layer</span>
    <span class="n">adata_x</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span>  <span class="c1"># Main gene expression matrix</span>

    <span class="c1"># Align gene symbols between the marker dataframe and the AnnData object</span>
    <span class="n">genes</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>  <span class="c1"># List of genes in AnnData</span>
    <span class="n">marker_genes</span> <span class="o">=</span> <span class="n">marker_df</span><span class="p">[</span><span class="s1">&#39;gene_symbol&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>  <span class="c1"># List of marker genes</span>
    <span class="n">common_genes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">marker_genes</span><span class="p">)</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">genes</span><span class="p">))</span>  <span class="c1"># Find common genes</span>

    <span class="c1"># Filter the marker DataFrame to only include the common genes</span>
    <span class="n">marker_df</span> <span class="o">=</span> <span class="n">marker_df</span><span class="p">[</span><span class="n">marker_df</span><span class="p">[</span><span class="s1">&#39;gene_symbol&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">common_genes</span><span class="p">)]</span>

    <span class="c1"># Initialize a list to store results for all cell types</span>
    <span class="n">all_results</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="c1"># Group marker genes by cell type</span>
    <span class="k">for</span> <span class="n">cell_type</span><span class="p">,</span> <span class="n">group</span> <span class="ow">in</span> <span class="n">marker_df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;cell_type&quot;</span><span class="p">):</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Processing cell_type: </span><span class="si">{</span><span class="n">cell_type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># Filter marker genes with p-value &lt; 0.05</span>
        <span class="n">filtered_group</span> <span class="o">=</span> <span class="n">group</span><span class="p">[</span><span class="n">group</span><span class="p">[</span><span class="s1">&#39;pvalue&#39;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mf">0.05</span><span class="p">]</span>

        <span class="c1"># If more than 50 genes are available, select the top 50 based on the &#39;score&#39; column</span>
        <span class="n">top_genes</span> <span class="o">=</span> <span class="n">filtered_group</span><span class="o">.</span><span class="n">nlargest</span><span class="p">(</span><span class="mi">50</span><span class="p">,</span> <span class="s2">&quot;score&quot;</span><span class="p">)</span>

        <span class="c1"># Define a helper function to compute cosine similarity</span>
        <span class="k">def</span> <span class="nf">compute_cosine_similarity</span><span class="p">(</span><span class="n">gene_symbol</span><span class="p">,</span> <span class="n">mapping_ori</span><span class="p">,</span> <span class="n">adata_x</span><span class="p">,</span> <span class="n">var_names</span><span class="p">):</span>
<span class="w">            </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">            Compute cosine similarity for a given gene.</span>

<span class="sd">            Parameters:</span>
<span class="sd">            ----------</span>
<span class="sd">            gene_symbol : str</span>
<span class="sd">                The gene symbol to compute similarity for.</span>
<span class="sd">            mapping_ori : np.ndarray</span>
<span class="sd">                The original mapping layer array.</span>
<span class="sd">            adata_x : np.ndarray</span>
<span class="sd">                Gene expression matrix from AnnData.</span>
<span class="sd">            var_names : list</span>
<span class="sd">                List of gene names in AnnData.</span>

<span class="sd">            Returns:</span>
<span class="sd">            -------</span>
<span class="sd">            float</span>
<span class="sd">                Cosine similarity value or NaN if computation is not possible.</span>
<span class="sd">            &quot;&quot;&quot;</span>
            <span class="c1"># If the gene is not found in AnnData, return NaN</span>
            <span class="k">if</span> <span class="n">gene_symbol</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">var_names</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>

            <span class="c1"># Get the index of the gene in the AnnData object</span>
            <span class="n">gene_idx</span> <span class="o">=</span> <span class="n">var_names</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">gene_symbol</span><span class="p">)</span>

            <span class="c1"># Extract expression vectors for the gene</span>
            <span class="n">gene_expression_vector</span> <span class="o">=</span> <span class="n">adata_x</span><span class="p">[:,</span> <span class="n">gene_idx</span><span class="p">]</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>  <span class="c1"># Expression vector in adata.X</span>
            <span class="n">ori_vector</span> <span class="o">=</span> <span class="n">mapping_ori</span><span class="p">[:,</span> <span class="n">gene_idx</span><span class="p">]</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>  <span class="c1"># Expression vector in mapping_ori</span>

            <span class="c1"># Compute cosine similarity if vectors are non-zero</span>
            <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">gene_expression_vector</span><span class="p">)</span> <span class="ow">and</span> <span class="n">np</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">ori_vector</span><span class="p">):</span>
                <span class="n">similarity</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">-</span> <span class="n">cosine</span><span class="p">(</span><span class="n">gene_expression_vector</span><span class="p">,</span> <span class="n">ori_vector</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">similarity</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>  <span class="c1"># Return NaN for zero vectors</span>

            <span class="k">return</span> <span class="n">similarity</span>

        <span class="c1"># Compute cosine similarity for each gene in the top genes</span>
        <span class="n">var_names</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>  <span class="c1"># Get gene names from AnnData</span>
        <span class="n">cosine_similarities</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">compute_cosine_similarity</span><span class="p">(</span><span class="n">gene_symbol</span><span class="p">,</span> <span class="n">mapping_ori</span><span class="p">,</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> <span class="n">var_names</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">gene_symbol</span> <span class="ow">in</span> <span class="n">top_genes</span><span class="p">[</span><span class="s2">&quot;gene_symbol&quot;</span><span class="p">]</span>
        <span class="p">]</span>

        <span class="c1"># Add cosine similarity values to the DataFrame</span>
        <span class="n">top_genes</span><span class="p">[</span><span class="s2">&quot;cosine_similarity&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cosine_similarities</span>

        <span class="c1"># Append the results for the current cell type</span>
        <span class="n">all_results</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">top_genes</span><span class="p">)</span>

    <span class="c1"># Concatenate results from all cell types into a single DataFrame</span>
    <span class="n">final_results</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">all_results</span><span class="p">,</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># Use seaborn&#39;s default palette if no custom palette is provided</span>
  <span class="c1"># Use seaborn&#39;s default color palette</span>

    <span class="c1"># Create a Swarm Plot for cosine similarity grouped by cell type</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>  <span class="c1"># Set the figure size to a smaller size (6x4 inches)</span>
    <span class="n">sns</span><span class="o">.</span><span class="n">swarmplot</span><span class="p">(</span>
        <span class="n">x</span><span class="o">=</span><span class="s2">&quot;cell_type&quot;</span><span class="p">,</span> 
        <span class="n">y</span><span class="o">=</span><span class="s2">&quot;cosine_similarity&quot;</span><span class="p">,</span> 
        <span class="n">data</span><span class="o">=</span><span class="n">final_results</span><span class="p">,</span> 
        <span class="n">palette</span><span class="o">=</span><span class="n">custom_palette</span>  <span class="c1"># Use custom or default palette</span>
    <span class="p">)</span>

    <span class="c1"># Add labels and customize the plot</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Cell Type&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Cosine Similarity&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>  <span class="c1"># Adjust font size for x-axis labels</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>  <span class="c1"># Ensure the layout fits well</span>

    <span class="c1"># Return the matplotlib.pyplot object for further customization or saving</span>
    <span class="k">return</span> <span class="n">plt</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div><hr />
<p><div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">plot_he_cell_type</span><span class="p">(</span><span class="n">data</span><span class="p">,</span><span class="n">out_dir</span><span class="p">)</span>
</code></pre></div>
</p>


<div class="doc doc-object doc-function">



<a id="cytobulk.plots._plot.plot_he_cell_type"></a>
    <div class="doc doc-contents first">

        <p>Visualize and save a scatter plot of cell locations by cell type.</p>
<p>This function generates a scatter plot to visualize the spatial distribution of different cell types 
based on their x and y coordinates. The plot is customized with specific colors for predefined cell types 
and is saved as a PNG file.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>data</code>
            </td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A pandas DataFrame containing the following columns:
- 'cell_type': Categorical data representing cell types (e.g., "Epithelial Cells").
- 'x': Numerical values representing the x-coordinates of the cells.
- 'y': Numerical values representing the y-coordinates of the cells.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>out_dir</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The directory where the generated plot image will be saved. The plot will be saved as 
"cell_type.png" within this directory.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>None</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>This function does not return any values. It generates and saves the scatter plot as a PNG file.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<details class="note" open>
  <summary>Notes</summary>
  <ul>
<li>The plot uses predefined colors for the following cell types:<ul>
<li>"Epithelial Cells": #A52A2A (brown)</li>
<li>"Neutrophils": #0000B8 (blue)</li>
<li>"Plasma Cells": #0D98BA (cyan)</li>
<li>"Connective Tissue": #FFCC33 (yellow)</li>
<li>"Lymphocytes": #B284BE (purple)
  Cell types not listed above are assigned the color "black".</li>
</ul>
</li>
<li>The y-axis is inverted to match the typical orientation of spatial data.</li>
</ul>
</details>
            <details class="quote">
              <summary>Source code in <code>cytobulk\plots\_plot.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">plot_he_cell_type</span><span class="p">(</span><span class="n">data</span><span class="p">,</span><span class="n">out_dir</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Visualize and save a scatter plot of cell locations by cell type.</span>

<span class="sd">    This function generates a scatter plot to visualize the spatial distribution of different cell types </span>
<span class="sd">    based on their x and y coordinates. The plot is customized with specific colors for predefined cell types </span>
<span class="sd">    and is saved as a PNG file.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    data : pandas.DataFrame</span>
<span class="sd">        A pandas DataFrame containing the following columns:</span>
<span class="sd">        - &#39;cell_type&#39;: Categorical data representing cell types (e.g., &quot;Epithelial Cells&quot;).</span>
<span class="sd">        - &#39;x&#39;: Numerical values representing the x-coordinates of the cells.</span>
<span class="sd">        - &#39;y&#39;: Numerical values representing the y-coordinates of the cells.</span>

<span class="sd">    out_dir : str</span>
<span class="sd">        The directory where the generated plot image will be saved. The plot will be saved as </span>
<span class="sd">        &quot;cell_type.png&quot; within this directory.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    None</span>
<span class="sd">        This function does not return any values. It generates and saves the scatter plot as a PNG file.</span>

<span class="sd">    Notes</span>
<span class="sd">    -----</span>
<span class="sd">    - The plot uses predefined colors for the following cell types:</span>
<span class="sd">        * &quot;Epithelial Cells&quot;: #A52A2A (brown)</span>
<span class="sd">        * &quot;Neutrophils&quot;: #0000B8 (blue)</span>
<span class="sd">        * &quot;Plasma Cells&quot;: #0D98BA (cyan)</span>
<span class="sd">        * &quot;Connective Tissue&quot;: #FFCC33 (yellow)</span>
<span class="sd">        * &quot;Lymphocytes&quot;: #B284BE (purple)</span>
<span class="sd">      Cell types not listed above are assigned the color &quot;black&quot;.</span>
<span class="sd">    - The y-axis is inverted to match the typical orientation of spatial data.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
    <span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>

    <span class="n">categories</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
    <span class="n">colors</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;Epithelial Cells&quot;</span><span class="p">:</span> <span class="s2">&quot;#A52A2A&quot;</span><span class="p">,</span>  
        <span class="s2">&quot;Neutrophils&quot;</span><span class="p">:</span> <span class="s2">&quot;#0000B8&quot;</span><span class="p">,</span>      
        <span class="s2">&quot;Plasma Cells&quot;</span><span class="p">:</span> <span class="s2">&quot;#0D98BA&quot;</span><span class="p">,</span>     
        <span class="s2">&quot;Connective Tissue&quot;</span><span class="p">:</span> <span class="s2">&quot;#FFCC33&quot;</span><span class="p">,</span> 
        <span class="s2">&quot;Lymphocytes&quot;</span><span class="p">:</span> <span class="s2">&quot;#B284BE&quot;</span>      
    <span class="p">}</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>

    <span class="k">for</span> <span class="n">category</span> <span class="ow">in</span> <span class="n">categories</span><span class="p">:</span>
        <span class="n">subset</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">category</span><span class="p">]</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">subset</span><span class="p">[</span><span class="s2">&quot;x&quot;</span><span class="p">],</span> <span class="n">subset</span><span class="p">[</span><span class="s2">&quot;y&quot;</span><span class="p">],</span> <span class="n">label</span><span class="o">=</span><span class="n">category</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">category</span><span class="p">,</span> <span class="s2">&quot;black&quot;</span><span class="p">),</span> <span class="n">s</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.8</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span>
        <span class="n">loc</span><span class="o">=</span><span class="s2">&quot;upper right&quot;</span><span class="p">,</span>             <span class="c1"># Place the legend in the top-right corner</span>
        <span class="n">title</span><span class="o">=</span><span class="s2">&quot;Cell Types&quot;</span><span class="p">,</span>            <span class="c1"># Add a title to the legend</span>
        <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span>                    <span class="c1"># Set font size for the legend labels</span>
        <span class="n">title_fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>             <span class="c1"># Set font size for the legend title</span>
        <span class="n">markerscale</span><span class="o">=</span><span class="mf">1.5</span>                <span class="c1"># Scale the legend markers (scatter points)</span>
    <span class="p">)</span>


    <span class="n">plt</span><span class="o">.</span><span class="n">gca</span><span class="p">()</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s2">&quot;off&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">gca</span><span class="p">()</span><span class="o">.</span><span class="n">invert_yaxis</span><span class="p">()</span>
    <span class="n">save_path</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">out_dir</span><span class="si">}</span><span class="s2">/cell_type.png&quot;</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">save_path</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s2">&quot;tight&quot;</span><span class="p">,</span> <span class="n">pad_inches</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span> 
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    
      
      <nav class="md-footer__inner md-grid" aria-label="Footer" >
        
          
          <a href="../../tools/_segmentation/" class="md-footer__link md-footer__link--prev" aria-label="Previous: segmentation">
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
            </div>
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Previous
              </span>
              <div class="md-ellipsis">
                segmentation
              </div>
            </div>
          </a>
        
        
          
          <a href="../../tutorials/mapping_sc_bulk/" class="md-footer__link md-footer__link--next" aria-label="Next: Bulk deconvolution and mapping with scRNA-seq">
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Next
              </span>
              <div class="md-ellipsis">
                Bulk deconvolution and mapping with scRNA-seq
              </div>
            </div>
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11z"/></svg>
            </div>
          </a>
        
      </nav>
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Copyright &copy; 2023 WANG Xueying WANG Yian
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["search.suggest", "search.highlight", "search.share\\", "navigation.footer"], "search": "../../assets/javascripts/workers/search.6ce7567c.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.88dd0f4e.min.js"></script>
      
    
  </body>
</html>