
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../../Install/">
      
      
        <link rel="next" href="../_mapping/">
      
      
      <link rel="icon" href="../../assets/favicon_temp.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.5.49">
    
    
      
        <title>deconv - CytoBulk's Docs</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.6f8fc17f.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="../../css/ansi-colours.css">
    
      <link rel="stylesheet" href="../../css/jupyter-cells.css">
    
      <link rel="stylesheet" href="../../css/pandas-dataframe.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#tools_deconv" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="CytoBulk&#39;s Docs" class="md-header__button md-logo" aria-label="CytoBulk's Docs" data-md-component="logo">
      
  <img src="../../assets/logo_temp.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            CytoBulk's Docs
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              deconv
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m14.3 16-.7-2h-3.2l-.7 2H7.8L11 7h2l3.2 9zM20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12zm-9.15 3.96h2.3L12 9z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_2" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 7a5 5 0 0 1 5 5 5 5 0 0 1-5 5 5 5 0 0 1-5-5 5 5 0 0 1 5-5m0 2a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0-7 2.39 3.42C13.65 5.15 12.84 5 12 5s-1.65.15-2.39.42zM3.34 7l4.16-.35A7.2 7.2 0 0 0 5.94 8.5c-.44.74-.69 1.5-.83 2.29zm.02 10 1.76-3.77a7.131 7.131 0 0 0 2.38 4.14zM20.65 7l-1.77 3.79a7.02 7.02 0 0 0-2.38-4.15zm-.01 10-4.14.36c.59-.51 1.12-1.14 1.54-1.86.42-.73.69-1.5.83-2.29zM12 22l-2.41-3.44c.74.27 1.55.44 2.41.44.82 0 1.63-.17 2.37-.44z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_2">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3zm3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95zm-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/kristaxying/CytoBulk/tutorials" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    CytoBulk
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="CytoBulk&#39;s Docs" class="md-nav__button md-logo" aria-label="CytoBulk's Docs" data-md-component="logo">
      
  <img src="../../assets/logo_temp.png" alt="logo">

    </a>
    CytoBulk's Docs
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/kristaxying/CytoBulk/tutorials" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    CytoBulk
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../Install/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Install
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" checked>
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    API References
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            API References
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    deconv
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    deconv
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#cytobulk.tools._deconv.bulk_deconv" class="md-nav__link">
    <span class="md-ellipsis">
      bulk_deconv
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#cytobulk.tools._deconv.st_deconv" class="md-nav__link">
    <span class="md-ellipsis">
      st_deconv
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../_mapping/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    mapping
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../_segmentation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    segmentation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../plots/_plot/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    plot
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Examples
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Examples
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tutorials/mapping_sc_bulk/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Bulk deconvolution and mapping with scRNA-seq
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tutorials/mapping_sc_st/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ST deconvolution and mapping with scRNA-seq
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tutorials/mapping_he_sc/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    H&E cell prediction and integration with scRNA-seq
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../References/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    References
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#cytobulk.tools._deconv.bulk_deconv" class="md-nav__link">
    <span class="md-ellipsis">
      bulk_deconv
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#cytobulk.tools._deconv.st_deconv" class="md-nav__link">
    <span class="md-ellipsis">
      st_deconv
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


<h1 id="tools_deconv">tools._deconv</h1>
<hr />
<p><div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">bulk_deconv</span><span class="p">(</span><span class="n">bulk_data</span><span class="p">,</span>
                <span class="n">sc_adata</span><span class="p">,</span>
                <span class="n">annotation_key</span><span class="p">,</span>
                <span class="n">marker_data</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                <span class="n">rename</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                <span class="n">dataset_name</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
                <span class="n">out_dir</span><span class="o">=</span><span class="s1">&#39;.&#39;</span><span class="p">,</span>
                <span class="n">different_source</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">cell_list</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                <span class="n">scale_factors</span><span class="o">=</span><span class="mi">10000</span><span class="p">,</span>
                <span class="n">trans_method</span><span class="o">=</span><span class="s2">&quot;log&quot;</span><span class="p">,</span>
                <span class="n">save</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
                <span class="n">save_figure</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">n_cell</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span>
                <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
</code></pre></div>
</p>


<div class="doc doc-object doc-function">



<a id="cytobulk.tools._deconv.bulk_deconv"></a>
    <div class="doc doc-contents first">

        <p>Deconvolute the cell type fraction from bulk expression data with single cell dataset as reference.
Reconstruct the bulk data using single cell.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>bulk_data</code>
            </td>
            <td>
                  <code>dataframe</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>An :class:<code>~pandas.dataframe</code> containing the bulk expression data. 
The first column should be gene symbol, following column should be sample name.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>sc_adata</code>
            </td>
            <td>
                  <code><span title="anndata.AnnData">AnnData</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>An :class:<code>~anndata.AnnData</code> containing the single cell expression.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>annotation_key</code>
            </td>
            <td>
                  <code>string</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The <code>.obs</code> key where the single cell annotation is stored. : anndata.AnnData.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>marker_data</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>An :class:<code>~pandas.dataframe</code> which columns are cell types, rows are marker gene.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>dataset_name</code>
            </td>
            <td>
                  <code>string.</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The prefix of output file.</p>
              </div>
            </td>
            <td>
                  <code>&#39;&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>out_dir</code>
            </td>
            <td>
                  <code>string</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The path to store the output data.</p>
              </div>
            </td>
            <td>
                  <code>&#39;.&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>different_source</code>
            </td>
            <td>
                  <code>boolean</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>True for single cell and bulk data from the same sample, which means not executing batch effect.
False for single cell and bulk data from the different samples, which means executing batch effect.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>cell_list</code>
            </td>
            <td>
                  <code>list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The list indicate the cell type names which need to take into consideration.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>scale_factors</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The number of counts to normalize every observation to before computing profiles. If <code>None</code>, no normalization is performed.</p>
              </div>
            </td>
            <td>
                  <code>100000</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>trans_method</code>
            </td>
            <td>
                  <code>string</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>What transformation to apply to the expression before computing the profiles. 
- "log" : log(x+1)
- <code>None</code> : no transformation</p>
              </div>
            </td>
            <td>
                  <code>&#39;log&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>save</code>
            </td>
            <td>
                  <code>boolean</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether save the result data during each step. If saving, the processing may be skipped.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>save_figure</code>
            </td>
            <td>
                  <code>boolean</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether save figures during preprocessing. eg. scatter plot of pca data.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>mapping_sc</code>
            </td>
            <td>
                  <code>boolean</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether reconstruct the bulk data with single cell data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>n_cell</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The number of cells within each bulk.</p>
              </div>
            </td>
            <td>
                  <code>2000</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>**kwargs</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Additional keyword arguments forwarded to
:func:<code>~cytobulk.preprocessing.qc_bulk_sc</code>.</p>
              </div>
            </td>
            <td>
                  <code>{}</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>Returns the deconvolution result and reconstructed bulk.</code>
            </td>
            <td>
              <div class="doc-md-description">
                
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>cytobulk\tools\_deconv.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">bulk_deconv</span><span class="p">(</span><span class="n">bulk_data</span><span class="p">,</span>
                <span class="n">sc_adata</span><span class="p">,</span>
                <span class="n">annotation_key</span><span class="p">,</span>
                <span class="n">marker_list</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                <span class="n">rename</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                <span class="n">dataset_name</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
                <span class="n">out_dir</span><span class="o">=</span><span class="s1">&#39;.&#39;</span><span class="p">,</span>
                <span class="n">different_source</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">cell_list</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                <span class="n">scale_factors</span><span class="o">=</span><span class="mi">100000</span><span class="p">,</span>
                <span class="n">trans_method</span><span class="o">=</span><span class="s2">&quot;log&quot;</span><span class="p">,</span>
                <span class="n">save</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
                <span class="n">save_figure</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">n_cell</span><span class="o">=</span><span class="mi">2000</span><span class="p">,</span>
                <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Deconvolute the cell type fraction from bulk expression data with single cell dataset as reference.</span>
<span class="sd">    Reconstruct the bulk data using single cell.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    bulk_data : dataframe</span>
<span class="sd">        An :class:`~pandas.dataframe` containing the bulk expression data. </span>
<span class="sd">        The first column should be gene symbol, following column should be sample name.</span>
<span class="sd">    sc_adata : anndata.AnnData</span>
<span class="sd">        An :class:`~anndata.AnnData` containing the single cell expression.</span>
<span class="sd">    annotation_key : string</span>
<span class="sd">        The `.obs` key where the single cell annotation is stored. : anndata.AnnData.</span>
<span class="sd">    marker_data : </span>
<span class="sd">        An :class:`~pandas.dataframe` which columns are cell types, rows are marker gene.</span>
<span class="sd">    dataset_name : string.</span>
<span class="sd">        The prefix of output file.</span>
<span class="sd">    out_dir : string, optional</span>
<span class="sd">        The path to store the output data.</span>
<span class="sd">    different_source : boolean, optional</span>
<span class="sd">        True for single cell and bulk data from the same sample, which means not executing batch effect.</span>
<span class="sd">        False for single cell and bulk data from the different samples, which means executing batch effect.</span>
<span class="sd">    cell_list : list, optional</span>
<span class="sd">        The list indicate the cell type names which need to take into consideration.</span>
<span class="sd">    scale_factors : int, optional</span>
<span class="sd">        The number of counts to normalize every observation to before computing profiles. If `None`, no normalization is performed. </span>
<span class="sd">    trans_method : string, optional</span>
<span class="sd">        What transformation to apply to the expression before computing the profiles. </span>
<span class="sd">        - &quot;log&quot; : log(x+1)</span>
<span class="sd">        - `None` : no transformation</span>
<span class="sd">    save : boolean, optional</span>
<span class="sd">        Whether save the result data during each step. If saving, the processing may be skipped.</span>
<span class="sd">    save_figure : boolean, optional</span>
<span class="sd">        Whether save figures during preprocessing. eg. scatter plot of pca data.</span>
<span class="sd">    mapping_sc : boolean, optional</span>
<span class="sd">        Whether reconstruct the bulk data with single cell data.</span>
<span class="sd">    n_cell : int, optional</span>
<span class="sd">        The number of cells within each bulk.</span>
<span class="sd">    **kwargs : </span>
<span class="sd">        Additional keyword arguments forwarded to</span>
<span class="sd">        :func:`~cytobulk.preprocessing.qc_bulk_sc`.</span>


<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    Returns the deconvolution result and reconstructed bulk.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># check the filtered dataset. If exist, skipping preprocessing.</span>
    <span class="n">bulk_ori_adata</span> <span class="o">=</span> <span class="n">bulk_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">exists</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">out_dir</span><span class="si">}</span><span class="s1">/filtered/pseudo_bulk_</span><span class="si">{</span><span class="n">dataset_name</span><span class="si">}</span><span class="s1">.h5ad&#39;</span><span class="p">)</span> <span class="ow">and</span> <span class="n">exists</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">out_dir</span><span class="si">}</span><span class="s1">/filtered/sc_data_</span><span class="si">{</span><span class="n">dataset_name</span><span class="si">}</span><span class="s1">.h5ad&#39;</span><span class="p">)</span> <span class="ow">and</span> \
    <span class="n">exists</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">out_dir</span><span class="si">}</span><span class="s1">/filtered/bulk_data_</span><span class="si">{</span><span class="n">dataset_name</span><span class="si">}</span><span class="s1">.h5ad&#39;</span><span class="p">)</span> <span class="ow">and</span> <span class="n">exists</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">out_dir</span><span class="si">}</span><span class="s1">/filtered/cells_</span><span class="si">{</span><span class="n">dataset_name</span><span class="si">}</span><span class="s1">.json&#39;</span><span class="p">):</span>

        <span class="n">pseudo_bulk</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">out_dir</span><span class="si">}</span><span class="s2">/filtered/pseudo_bulk_</span><span class="si">{</span><span class="n">dataset_name</span><span class="si">}</span><span class="s2">.h5ad&quot;</span><span class="p">)</span>
        <span class="n">sc_adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">out_dir</span><span class="si">}</span><span class="s2">/filtered/sc_data_</span><span class="si">{</span><span class="n">dataset_name</span><span class="si">}</span><span class="s2">.h5ad&quot;</span><span class="p">)</span>
        <span class="n">bulk_adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">out_dir</span><span class="si">}</span><span class="s2">/filtered/bulk_data_</span><span class="si">{</span><span class="n">dataset_name</span><span class="si">}</span><span class="s2">.h5ad&quot;</span><span class="p">)</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">out_dir</span><span class="si">}</span><span class="s2">/filtered/cells_</span><span class="si">{</span><span class="n">dataset_name</span><span class="si">}</span><span class="s2">.json&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">json_file</span><span class="p">:</span>
            <span class="n">common_cell</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">json_file</span><span class="p">)</span>
        <span class="n">annotation_key</span><span class="o">=</span><span class="s2">&quot;curated_cell_type&quot;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1">#preprocessing</span>
        <span class="n">sc_adata</span><span class="p">,</span> <span class="n">pseudo_bulk</span><span class="p">,</span> <span class="n">bulk_adata</span><span class="p">,</span> <span class="n">common_cell</span><span class="p">,</span><span class="n">annotation_key</span> <span class="o">=</span> <span class="n">pp</span><span class="o">.</span><span class="n">preprocessing</span><span class="p">(</span><span class="n">bulk_data</span><span class="p">,</span>
                                                                        <span class="n">sc_adata</span><span class="p">,</span>
                                                                        <span class="n">annotation_key</span><span class="p">,</span>
                                                                        <span class="n">is_st</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                                                                        <span class="n">marker_list</span><span class="o">=</span><span class="n">marker_list</span><span class="p">,</span>
                                                                        <span class="n">rename</span><span class="o">=</span><span class="n">rename</span><span class="p">,</span>
                                                                        <span class="n">dataset_name</span><span class="o">=</span><span class="n">dataset_name</span><span class="p">,</span>
                                                                        <span class="n">out_dir</span><span class="o">=</span><span class="n">out_dir</span><span class="p">,</span>
                                                                        <span class="n">different_source</span><span class="o">=</span><span class="n">different_source</span><span class="p">,</span>
                                                                        <span class="n">cell_list</span><span class="o">=</span><span class="n">cell_list</span><span class="p">,</span>
                                                                        <span class="n">scale_factors</span><span class="o">=</span><span class="n">scale_factors</span><span class="p">,</span>
                                                                        <span class="n">trans_method</span><span class="o">=</span><span class="n">trans_method</span><span class="p">,</span>
                                                                        <span class="n">n_sample_each_group</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">bulk_data</span><span class="o">.</span><span class="n">obs_names</span><span class="p">)</span><span class="o">*</span><span class="mi">5</span><span class="p">,</span>
                                                                        <span class="n">min_cells_each_group</span><span class="o">=</span><span class="n">n_cell</span><span class="p">,</span>
                                                                        <span class="n">cell_gap_each_group</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span>
                                                                        <span class="n">group_number</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
                                                                        <span class="n">save</span> <span class="o">=</span> <span class="n">save</span><span class="p">,</span>
                                                                        <span class="n">save_figure</span><span class="o">=</span><span class="n">save_figure</span><span class="p">,</span>
                                                                        <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
    <span class="c1">#deconvolution</span>
    <span class="k">if</span> <span class="n">exists</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">out_dir</span><span class="si">}</span><span class="s1">/output/</span><span class="si">{</span><span class="n">dataset_name</span><span class="si">}</span><span class="s1">_prediction_frac.csv&#39;</span><span class="p">):</span>
        <span class="n">deconv_result</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">out_dir</span><span class="si">}</span><span class="s1">/output/</span><span class="si">{</span><span class="n">dataset_name</span><span class="si">}</span><span class="s1">_prediction_frac.csv&#39;</span><span class="p">,</span><span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">deconv_result</span> <span class="o">=</span> <span class="n">_bulk_sc_deconv</span><span class="p">(</span><span class="n">bulk_adata</span><span class="p">,</span> 
                                        <span class="n">pseudo_bulk</span><span class="p">,</span> 
                                        <span class="n">sc_adata</span><span class="p">,</span> 
                                        <span class="n">common_cell</span><span class="p">,</span>
                                        <span class="n">annotation_key</span> <span class="o">=</span> <span class="n">annotation_key</span><span class="p">,</span> 
                                        <span class="n">dataset_name</span><span class="o">=</span><span class="n">dataset_name</span><span class="p">,</span> 
                                        <span class="n">out_dir</span><span class="o">=</span><span class="n">out_dir</span><span class="p">)</span>
    <span class="n">row_sums</span> <span class="o">=</span> <span class="n">deconv_result</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">df_normalized</span> <span class="o">=</span> <span class="n">deconv_result</span><span class="o">.</span><span class="n">div</span><span class="p">(</span><span class="n">row_sums</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">bulk_ori_adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;deconv&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">df_normalized</span>
    <span class="n">df_normalized</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">out_dir</span><span class="si">}</span><span class="s2">/output/</span><span class="si">{</span><span class="n">dataset_name</span><span class="si">}</span><span class="s2">_prediction_frac_normalized.csv&quot;</span><span class="p">)</span>
    <span class="n">bulk_ori_adata</span><span class="o">.</span><span class="n">write_h5ad</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">out_dir</span><span class="si">}</span><span class="s1">/output/</span><span class="si">{</span><span class="n">dataset_name</span><span class="si">}</span><span class="s1">_bulk_adata.h5ad&#39;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">deconv_result</span><span class="p">,</span><span class="n">bulk_adata</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div><hr />
<p><div class="highlight"><pre><span></span><code><span class="n">st_deconv</span><span class="p">(</span><span class="n">st_adata</span><span class="p">,</span>
            <span class="n">sc_adata</span><span class="p">,</span>
            <span class="n">annotation_key</span><span class="p">,</span>
            <span class="n">marker_list</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="n">rename</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="n">dataset_name</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
            <span class="n">out_dir</span><span class="o">=</span><span class="s1">&#39;.&#39;</span><span class="p">,</span>
            <span class="n">different_source</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">cell_list</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="n">scale_factors</span><span class="o">=</span><span class="mi">10000</span><span class="p">,</span>
            <span class="n">trans_method</span><span class="o">=</span><span class="s2">&quot;log&quot;</span><span class="p">,</span>
            <span class="n">save</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
            <span class="n">save_figure</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">n_cell</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
            <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
</code></pre></div>
</p>


<div class="doc doc-object doc-function">



<a id="cytobulk.tools._deconv.st_deconv"></a>
    <div class="doc doc-contents first">

        <p>Deconvolute the cell type fraction from spot expression data with single cell dataset as reference.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>st_adata</code>
            </td>
            <td>
                  <code><span title="anndata.AnnData">AnnData</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>An :class:<code>~anndata.AnnData</code> containing the single cell expression.
The first column should be gene symbol, following column should be spot name.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>sc_adata</code>
            </td>
            <td>
                  <code><span title="anndata.AnnData">AnnData</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>An :class:<code>~anndata.AnnData</code> containing the single cell expression.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>annotation_key</code>
            </td>
            <td>
                  <code>string</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The <code>.obs</code> key where the single cell annotation is stored. : anndata.AnnData.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>marker_list</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>An :class:<code>~pandas.dataframe</code> which columns are cell types, rows are marker gene.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>dataset_name</code>
            </td>
            <td>
                  <code>string.</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The prefix of output file.</p>
              </div>
            </td>
            <td>
                  <code>&#39;&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>out_dir</code>
            </td>
            <td>
                  <code>string</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The path to store the output data.</p>
              </div>
            </td>
            <td>
                  <code>&#39;.&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>different_source</code>
            </td>
            <td>
                  <code>boolean</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>True for single cell and bulk data from the same sample, which means not executing batch effect.
False for single cell and bulk data from the different samples, which means executing batch effect.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>cell_list</code>
            </td>
            <td>
                  <code>list</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The list indicate the cell type names which need to take into consideration.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>scale_factors</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The number of counts to normalize every observation to before computing profiles. If <code>None</code>, no normalization is performed.</p>
              </div>
            </td>
            <td>
                  <code>10000</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>trans_method</code>
            </td>
            <td>
                  <code>string</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>What transformation to apply to the expression before computing the profiles. 
- "log" : log(x+1)
- <code>None</code> : no transformation</p>
              </div>
            </td>
            <td>
                  <code>&#39;log&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>save</code>
            </td>
            <td>
                  <code>boolean</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether save the result data during each step. If saving, the processing may be skipped.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>save_figure</code>
            </td>
            <td>
                  <code>boolean</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether save figures during preprocessing. eg. scatter plot of pca data.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>n_cell</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The number of cells within each bulk.</p>
              </div>
            </td>
            <td>
                  <code>10</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>**kwargs</code>
            </td>
            <td>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Additional keyword arguments forwarded to
:func:<code>~cytobulk.preprocessing.qc_bulk_sc</code>.</p>
              </div>
            </td>
            <td>
                  <code>{}</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>Returns the deconvolution result and reconstructed st.</code>
            </td>
            <td>
              <div class="doc-md-description">
                
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>cytobulk\tools\_deconv.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">st_deconv</span><span class="p">(</span><span class="n">st_adata</span><span class="p">,</span>
            <span class="n">sc_adata</span><span class="p">,</span>
            <span class="n">annotation_key</span><span class="p">,</span>
            <span class="n">marker_list</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="n">rename</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="n">dataset_name</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
            <span class="n">out_dir</span><span class="o">=</span><span class="s1">&#39;.&#39;</span><span class="p">,</span>
            <span class="n">different_source</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">cell_list</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="n">scale_factors</span><span class="o">=</span><span class="mi">10000</span><span class="p">,</span>
            <span class="n">trans_method</span><span class="o">=</span><span class="s2">&quot;log&quot;</span><span class="p">,</span>
            <span class="n">save</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
            <span class="n">save_figure</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">n_cell</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
            <span class="n">max_cells</span><span class="o">=</span><span class="mi">40</span><span class="p">,</span>
            <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Deconvolute the cell type fraction from spot expression data with single cell dataset as reference.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    st_adata : anndata.AnnData</span>
<span class="sd">        An :class:`~anndata.AnnData` containing the single cell expression.</span>
<span class="sd">        The first column should be gene symbol, following column should be spot name.</span>
<span class="sd">    sc_adata : anndata.AnnData</span>
<span class="sd">        An :class:`~anndata.AnnData` containing the single cell expression.</span>
<span class="sd">    annotation_key : string</span>
<span class="sd">        The `.obs` key where the single cell annotation is stored. : anndata.AnnData.</span>
<span class="sd">    marker_list : </span>
<span class="sd">        An :class:`~pandas.dataframe` which columns are cell types, rows are marker gene.</span>
<span class="sd">    dataset_name : string.</span>
<span class="sd">        The prefix of output file.</span>
<span class="sd">    out_dir : string, optional</span>
<span class="sd">        The path to store the output data.</span>
<span class="sd">    different_source : boolean, optional</span>
<span class="sd">        True for single cell and bulk data from the same sample, which means not executing batch effect.</span>
<span class="sd">        False for single cell and bulk data from the different samples, which means executing batch effect.</span>
<span class="sd">    cell_list : list, optional</span>
<span class="sd">        The list indicate the cell type names which need to take into consideration.</span>
<span class="sd">    scale_factors : int, optional</span>
<span class="sd">        The number of counts to normalize every observation to before computing profiles. If `None`, no normalization is performed. </span>
<span class="sd">    trans_method : string, optional</span>
<span class="sd">        What transformation to apply to the expression before computing the profiles. </span>
<span class="sd">        - &quot;log&quot; : log(x+1)</span>
<span class="sd">        - `None` : no transformation</span>
<span class="sd">    save : boolean, optional</span>
<span class="sd">        Whether save the result data during each step. If saving, the processing may be skipped.</span>
<span class="sd">    save_figure : boolean, optional</span>
<span class="sd">        Whether save figures during preprocessing. eg. scatter plot of pca data.</span>
<span class="sd">    n_cell : int, optional</span>
<span class="sd">        The number of cells within each bulk.</span>
<span class="sd">    **kwargs : </span>
<span class="sd">        Additional keyword arguments forwarded to</span>
<span class="sd">        :func:`~cytobulk.preprocessing.qc_bulk_sc`.</span>


<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    Returns the deconvolution result and reconstructed st.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># check the filtered dataset. If exist, skipping preprocessing.</span>
    <span class="n">st_ori_adata</span> <span class="o">=</span> <span class="n">st_adata</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">exists</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">out_dir</span><span class="si">}</span><span class="s1">/filtered/pseudo_bulk_</span><span class="si">{</span><span class="n">dataset_name</span><span class="si">}</span><span class="s1">.h5ad&#39;</span><span class="p">)</span> <span class="ow">and</span> <span class="n">exists</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">out_dir</span><span class="si">}</span><span class="s1">/filtered/sc_data_</span><span class="si">{</span><span class="n">dataset_name</span><span class="si">}</span><span class="s1">.h5ad&#39;</span><span class="p">)</span> <span class="ow">and</span> \
    <span class="n">exists</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">out_dir</span><span class="si">}</span><span class="s1">/filtered/bulk_data_</span><span class="si">{</span><span class="n">dataset_name</span><span class="si">}</span><span class="s1">.h5ad&#39;</span><span class="p">)</span> <span class="ow">and</span> <span class="n">exists</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">out_dir</span><span class="si">}</span><span class="s1">/filtered/cells_</span><span class="si">{</span><span class="n">dataset_name</span><span class="si">}</span><span class="s1">.json&#39;</span><span class="p">):</span>
        <span class="n">pseudo_st</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">out_dir</span><span class="si">}</span><span class="s2">/filtered/pseudo_bulk_</span><span class="si">{</span><span class="n">dataset_name</span><span class="si">}</span><span class="s2">.h5ad&quot;</span><span class="p">)</span>
        <span class="n">sc_adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">out_dir</span><span class="si">}</span><span class="s2">/filtered/sc_data_</span><span class="si">{</span><span class="n">dataset_name</span><span class="si">}</span><span class="s2">.h5ad&quot;</span><span class="p">)</span>
        <span class="n">st_adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">out_dir</span><span class="si">}</span><span class="s2">/filtered/bulk_data_</span><span class="si">{</span><span class="n">dataset_name</span><span class="si">}</span><span class="s2">.h5ad&quot;</span><span class="p">)</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">out_dir</span><span class="si">}</span><span class="s2">/filtered/cells_</span><span class="si">{</span><span class="n">dataset_name</span><span class="si">}</span><span class="s2">.json&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">json_file</span><span class="p">:</span>
            <span class="n">common_cell</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">json_file</span><span class="p">)</span>
        <span class="n">annotation_key</span><span class="o">=</span><span class="s2">&quot;curated_cell_type&quot;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1">#preprocessing</span>
        <span class="n">sc_adata</span><span class="p">,</span> <span class="n">pseudo_st</span><span class="p">,</span> <span class="n">st_adata</span><span class="p">,</span><span class="n">common_cell</span><span class="p">,</span><span class="n">annotation_key</span> <span class="o">=</span> <span class="n">pp</span><span class="o">.</span><span class="n">preprocessing</span><span class="p">(</span><span class="n">st_adata</span><span class="p">,</span>
                                                                        <span class="n">sc_adata</span><span class="p">,</span>
                                                                        <span class="n">annotation_key</span><span class="p">,</span>
                                                                        <span class="n">is_st</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                                                                        <span class="n">marker_list</span><span class="o">=</span><span class="n">marker_list</span><span class="p">,</span>
                                                                        <span class="n">rename</span><span class="o">=</span><span class="n">rename</span><span class="p">,</span>
                                                                        <span class="n">dataset_name</span><span class="o">=</span><span class="n">dataset_name</span><span class="p">,</span>
                                                                        <span class="n">out_dir</span><span class="o">=</span><span class="n">out_dir</span><span class="p">,</span>
                                                                        <span class="n">different_source</span><span class="o">=</span><span class="n">different_source</span><span class="p">,</span>
                                                                        <span class="n">cell_list</span><span class="o">=</span><span class="n">cell_list</span><span class="p">,</span>
                                                                        <span class="n">scale_factors</span><span class="o">=</span><span class="n">scale_factors</span><span class="p">,</span>
                                                                        <span class="n">trans_method</span><span class="o">=</span><span class="n">trans_method</span><span class="p">,</span>
                                                                        <span class="n">n_sample_each_group</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">st_adata</span><span class="o">.</span><span class="n">obs_names</span><span class="p">)</span><span class="o">*</span><span class="mi">6</span><span class="p">,</span>
                                                                        <span class="n">min_cells_each_group</span><span class="o">=</span><span class="n">n_cell</span><span class="p">,</span>
                                                                        <span class="n">cell_gap_each_group</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
                                                                        <span class="n">group_number</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
                                                                        <span class="n">save</span> <span class="o">=</span> <span class="n">save</span><span class="p">,</span>
                                                                        <span class="n">save_figure</span><span class="o">=</span><span class="n">save_figure</span><span class="p">,</span>
                                                                        <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
    <span class="c1">#deconvolution</span>
    <span class="k">if</span> <span class="n">exists</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">out_dir</span><span class="si">}</span><span class="s1">/output/</span><span class="si">{</span><span class="n">dataset_name</span><span class="si">}</span><span class="s1">_prediction_frac.csv&#39;</span><span class="p">):</span>
        <span class="n">deconv_result</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">out_dir</span><span class="si">}</span><span class="s1">/output/</span><span class="si">{</span><span class="n">dataset_name</span><span class="si">}</span><span class="s1">_prediction_frac.csv&#39;</span><span class="p">,</span><span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">deconv_result</span> <span class="o">=</span> <span class="n">_bulk_sc_deconv</span><span class="p">(</span><span class="n">st_adata</span><span class="p">,</span> 
                                        <span class="n">pseudo_st</span><span class="p">,</span> 
                                        <span class="n">sc_adata</span><span class="p">,</span> 
                                        <span class="n">common_cell</span><span class="p">,</span>
                                        <span class="n">annotation_key</span> <span class="o">=</span> <span class="n">annotation_key</span><span class="p">,</span> 
                                        <span class="n">dataset_name</span><span class="o">=</span><span class="n">dataset_name</span><span class="p">,</span> 
                                        <span class="n">out_dir</span><span class="o">=</span><span class="n">out_dir</span><span class="p">,</span>
                                        <span class="n">batch_effect</span><span class="o">=</span><span class="n">different_source</span><span class="p">,</span>
                                        <span class="n">is_st</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">threshold</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">/</span> <span class="n">max_cells</span>
    <span class="n">deconv_result</span><span class="p">[</span><span class="n">deconv_result</span> <span class="o">&lt;</span> <span class="n">threshold</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">row_sums</span> <span class="o">=</span> <span class="n">deconv_result</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">df_normalized</span> <span class="o">=</span> <span class="n">deconv_result</span><span class="o">.</span><span class="n">div</span><span class="p">(</span><span class="n">row_sums</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">st_ori_adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;deconv&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">df_normalized</span>
    <span class="n">df_normalized</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">out_dir</span><span class="si">}</span><span class="s2">/output/</span><span class="si">{</span><span class="n">dataset_name</span><span class="si">}</span><span class="s2">_prediction_frac_normalized.csv&quot;</span><span class="p">)</span>
    <span class="n">st_ori_adata</span><span class="o">.</span><span class="n">write_h5ad</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">out_dir</span><span class="si">}</span><span class="s1">/output/</span><span class="si">{</span><span class="n">dataset_name</span><span class="si">}</span><span class="s1">_st_adata.h5ad&#39;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">df_normalized</span><span class="p">,</span><span class="n">st_ori_adata</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    
      
      <nav class="md-footer__inner md-grid" aria-label="Footer" >
        
          
          <a href="../../Install/" class="md-footer__link md-footer__link--prev" aria-label="Previous: Install">
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
            </div>
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Previous
              </span>
              <div class="md-ellipsis">
                Install
              </div>
            </div>
          </a>
        
        
          
          <a href="../_mapping/" class="md-footer__link md-footer__link--next" aria-label="Next: mapping">
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Next
              </span>
              <div class="md-ellipsis">
                mapping
              </div>
            </div>
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11z"/></svg>
            </div>
          </a>
        
      </nav>
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Copyright &copy; 2023 WANG Xueying WANG Yian
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["search.suggest", "search.highlight", "search.share\\", "navigation.footer"], "search": "../../assets/javascripts/workers/search.6ce7567c.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.88dd0f4e.min.js"></script>
      
    
  </body>
</html>