
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../_deconv/">
      
      
        <link rel="next" href="../_segmentation/">
      
      
      <link rel="icon" href="../../assets/favicon_temp.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.5.49">
    
    
      
        <title>mapping - CytoBulk's Docs</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.6f8fc17f.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="../../css/ansi-colours.css">
    
      <link rel="stylesheet" href="../../css/jupyter-cells.css">
    
      <link rel="stylesheet" href="../../css/pandas-dataframe.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#tools_mapping" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="CytoBulk&#39;s Docs" class="md-header__button md-logo" aria-label="CytoBulk's Docs" data-md-component="logo">
      
  <img src="../../assets/logo_temp.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            CytoBulk's Docs
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              mapping
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m14.3 16-.7-2h-3.2l-.7 2H7.8L11 7h2l3.2 9zM20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12zm-9.15 3.96h2.3L12 9z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_2" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 7a5 5 0 0 1 5 5 5 5 0 0 1-5 5 5 5 0 0 1-5-5 5 5 0 0 1 5-5m0 2a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0-7 2.39 3.42C13.65 5.15 12.84 5 12 5s-1.65.15-2.39.42zM3.34 7l4.16-.35A7.2 7.2 0 0 0 5.94 8.5c-.44.74-.69 1.5-.83 2.29zm.02 10 1.76-3.77a7.131 7.131 0 0 0 2.38 4.14zM20.65 7l-1.77 3.79a7.02 7.02 0 0 0-2.38-4.15zm-.01 10-4.14.36c.59-.51 1.12-1.14 1.54-1.86.42-.73.69-1.5.83-2.29zM12 22l-2.41-3.44c.74.27 1.55.44 2.41.44.82 0 1.63-.17 2.37-.44z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_2">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3zm3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95zm-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://kristaxying.github.io/CytoBulk/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    CytoBulk
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="CytoBulk&#39;s Docs" class="md-nav__button md-logo" aria-label="CytoBulk's Docs" data-md-component="logo">
      
  <img src="../../assets/logo_temp.png" alt="logo">

    </a>
    CytoBulk's Docs
  </label>
  
    <div class="md-nav__source">
      <a href="https://kristaxying.github.io/CytoBulk/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    CytoBulk
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../Install/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Install
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" checked>
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    API References
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            API References
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../_deconv/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    deconv
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    mapping
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    mapping
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#cytobulk.tools._mapping.bulk_mapping" class="md-nav__link">
    <span class="md-ellipsis">
      bulk_mapping
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#cytobulk.tools._mapping.st_mapping" class="md-nav__link">
    <span class="md-ellipsis">
      st_mapping
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#cytobulk.tools._mapping.he_mapping" class="md-nav__link">
    <span class="md-ellipsis">
      he_mapping
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../_segmentation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    segmentation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../plots/_plot/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    plot
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Examples
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Examples
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tutorials/mapping_sc_bulk/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Bulk deconvolution and mapping with scRNA-seq
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tutorials/mapping_sc_st/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ST deconvolution and mapping with scRNA-seq
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../tutorials/mapping_he_sc/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    H&E cell prediction and integration with scRNA-seq
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../References/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    References
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#cytobulk.tools._mapping.bulk_mapping" class="md-nav__link">
    <span class="md-ellipsis">
      bulk_mapping
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#cytobulk.tools._mapping.st_mapping" class="md-nav__link">
    <span class="md-ellipsis">
      st_mapping
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#cytobulk.tools._mapping.he_mapping" class="md-nav__link">
    <span class="md-ellipsis">
      he_mapping
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


<h1 id="tools_mapping">tools._mapping</h1>
<hr />
<p><div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">bulk_mapping</span><span class="p">(</span><span class="n">frac_data</span><span class="p">,</span>
                <span class="n">bulk_adata</span><span class="p">,</span>
                <span class="n">sc_adata</span><span class="p">,</span>
                <span class="n">n_cell</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span>
                <span class="n">annotation_key</span><span class="o">=</span><span class="s2">&quot;curated_cell_type&quot;</span><span class="p">,</span>
                <span class="n">bulk_layer</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                <span class="n">sc_layer</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                <span class="n">reorder</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">multiprocessing</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">cpu_num</span><span class="o">=</span><span class="n">cpu_count</span><span class="p">()</span><span class="o">-</span><span class="mi">2</span><span class="p">,</span>
                <span class="n">dataset_name</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
                <span class="n">out_dir</span><span class="o">=</span><span class="s2">&quot;.&quot;</span><span class="p">,</span>
                <span class="n">normalization</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">filter_gene</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">cut_off_value</span><span class="o">=</span><span class="mf">0.6</span><span class="p">,</span>
                <span class="n">save</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</code></pre></div>
</p>


<div class="doc doc-object doc-function">



<a id="cytobulk.tools._mapping.bulk_mapping"></a>
    <div class="doc doc-contents first">

        <p>Reconstruct bulk data using single-cell data and cell type fractions.</p>
<p>This function maps bulk expression data to single-cell expression data using
cell type fraction information and various preprocessing steps.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>bulk_adata</code>
            </td>
            <td>
                  <code><span title="anndata.AnnData">AnnData</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>An :class:<code>~anndata.AnnData</code> object containing the input bulk data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>sc_adata</code>
            </td>
            <td>
                  <code><span title="anndata.AnnData">AnnData</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>An :class:<code>~anndata.AnnData</code> object containing the single-cell expression data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>n_cell</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of cells per bulk sample.</p>
              </div>
            </td>
            <td>
                  <code>100</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>annotation_key</code>
            </td>
            <td>
                  <code>string</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Key in <code>sc_adata.obs</code> for single-cell annotations.</p>
              </div>
            </td>
            <td>
                  <code>&#39;curated_cell_type&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>bulk_layer</code>
            </td>
            <td>
                  <code>string</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Layer in <code>bulk_adata</code> to use for bulk expression data.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>sc_layer</code>
            </td>
            <td>
                  <code>string</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Layer in <code>sc_adata</code> to use for single-cell expression data.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>reorder</code>
            </td>
            <td>
                  <code>bool, optional (default: True)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to reorder genes to ensure consistency between bulk and single-cell data.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>multiprocessing</code>
            </td>
            <td>
                  <code>bool, optional (default: True)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to use multiprocessing for efficiency.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>cpu_num</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of CPUs to use if multiprocessing is enabled.</p>
              </div>
            </td>
            <td>
                  <code><span title="multiprocessing.cpu_count">cpu_count</span>() - 4</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>project</code>
            </td>
            <td>
                  <code>string</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Prefix for output files.</p>
              </div>
            </td>
            <td>
                  <code>&#39;&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>out_dir</code>
            </td>
            <td>
                  <code>string</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Directory to store output files.</p>
              </div>
            </td>
            <td>
                  <code>&#39;.&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>normalization</code>
            </td>
            <td>
                  <code>bool, optional (default: True)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to apply CPM normalization to data.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>filter_gene</code>
            </td>
            <td>
                  <code>bool, optional (default: True)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to filter genes based on cosine similarity.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>cut_off_value</code>
            </td>
            <td>
                  <code>float, optional (default: 0.6)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Threshold for cosine similarity when filtering genes.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>save</code>
            </td>
            <td>
                  <code>bool, optional (default: True)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to save the result files.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>bulk_adata</code></td>            <td>
                  <code><span title="anndata.AnnData">AnnData</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The processed bulk data with mapping results.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>df</code></td>            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>DataFrame containing the mapping of bulk samples to single-cell IDs.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>cytobulk\tools\_mapping.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">bulk_mapping</span><span class="p">(</span><span class="n">bulk_adata</span><span class="p">,</span>
                <span class="n">sc_adata</span><span class="p">,</span>
                <span class="n">n_cell</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span>
                <span class="n">annotation_key</span><span class="o">=</span><span class="s2">&quot;curated_cell_type&quot;</span><span class="p">,</span>
                <span class="n">bulk_layer</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                <span class="n">sc_layer</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                <span class="n">reorder</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">multiprocessing</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">cpu_num</span><span class="o">=</span><span class="n">cpu_count</span><span class="p">()</span><span class="o">-</span><span class="mi">4</span><span class="p">,</span>
                <span class="n">project</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
                <span class="n">out_dir</span><span class="o">=</span><span class="s2">&quot;.&quot;</span><span class="p">,</span>
                <span class="n">normalization</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">filter_gene</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">save</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Reconstruct bulk data using single-cell data and cell type fractions.</span>

<span class="sd">    This function maps bulk expression data to single-cell expression data using</span>
<span class="sd">    cell type fraction information and various preprocessing steps.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    bulk_adata : anndata.AnnData</span>
<span class="sd">        An :class:`~anndata.AnnData` object containing the input bulk data.</span>

<span class="sd">    sc_adata : anndata.AnnData</span>
<span class="sd">        An :class:`~anndata.AnnData` object containing the single-cell expression data.</span>

<span class="sd">    n_cell : int, optional</span>
<span class="sd">        Number of cells per bulk sample.</span>

<span class="sd">    annotation_key : string, optional</span>
<span class="sd">        Key in `sc_adata.obs` for single-cell annotations.</span>

<span class="sd">    bulk_layer : string, optional</span>
<span class="sd">        Layer in `bulk_adata` to use for bulk expression data.</span>

<span class="sd">    sc_layer : string, optional</span>
<span class="sd">        Layer in `sc_adata` to use for single-cell expression data.</span>

<span class="sd">    reorder : bool, optional (default: True)</span>
<span class="sd">        Whether to reorder genes to ensure consistency between bulk and single-cell data.</span>

<span class="sd">    multiprocessing : bool, optional (default: True)</span>
<span class="sd">        Whether to use multiprocessing for efficiency.</span>

<span class="sd">    cpu_num : int, optional</span>
<span class="sd">        Number of CPUs to use if multiprocessing is enabled.</span>

<span class="sd">    project : string, optional</span>
<span class="sd">        Prefix for output files.</span>

<span class="sd">    out_dir : string, optional</span>
<span class="sd">        Directory to store output files.</span>

<span class="sd">    normalization : bool, optional (default: True)</span>
<span class="sd">        Whether to apply CPM normalization to data.</span>

<span class="sd">    filter_gene : bool, optional (default: True)</span>
<span class="sd">        Whether to filter genes based on cosine similarity.</span>

<span class="sd">    cut_off_value : float, optional (default: 0.6)</span>
<span class="sd">        Threshold for cosine similarity when filtering genes.</span>

<span class="sd">    save : bool, optional (default: True)</span>
<span class="sd">        Whether to save the result files.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    bulk_adata : anndata.AnnData</span>
<span class="sd">        The processed bulk data with mapping results.</span>

<span class="sd">    df : pandas.DataFrame</span>
<span class="sd">        DataFrame containing the mapping of bulk samples to single-cell IDs.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">start_t</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">perf_counter</span><span class="p">()</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;=================================================================================================&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Start to mapping bulk data with single cell dataset.&#39;</span><span class="p">)</span>
    <span class="c1"># format data</span>
    <span class="n">bulk_adata</span><span class="o">.</span><span class="n">var_names_make_unique</span><span class="p">()</span>
    <span class="n">sc_adata</span><span class="o">.</span><span class="n">var_names_make_unique</span><span class="p">()</span>
    <span class="n">intersect_gene</span> <span class="o">=</span> <span class="n">bulk_adata</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">sc_adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">)</span>
    <span class="n">bulk_adata</span> <span class="o">=</span> <span class="n">bulk_adata</span><span class="p">[:,</span><span class="n">intersect_gene</span><span class="p">]</span>
    <span class="n">sc_adata</span> <span class="o">=</span> <span class="n">sc_adata</span><span class="p">[:,</span><span class="n">intersect_gene</span><span class="p">]</span>
    <span class="n">cell_prop</span> <span class="o">=</span> <span class="n">bulk_adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;deconv&#39;</span><span class="p">]</span>
    <span class="n">cell_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">floor</span><span class="p">(</span><span class="n">n_cell</span> <span class="o">*</span> <span class="n">cell_prop</span><span class="p">)</span>
    <span class="n">cell_num</span> <span class="o">=</span> <span class="n">cell_matrix</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
    <span class="n">meta_data</span> <span class="o">=</span> <span class="n">sc_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[[</span><span class="n">annotation_key</span><span class="p">]]</span>
    <span class="n">meta_dict</span> <span class="o">=</span> <span class="n">meta_data</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">meta_data</span><span class="p">[</span><span class="n">annotation_key</span><span class="p">])</span><span class="o">.</span><span class="n">groups</span>
    <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">meta_dict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">meta_dict</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">value</span><span class="p">)</span>
    <span class="n">cellname_list</span><span class="o">=</span><span class="n">cell_prop</span><span class="o">.</span><span class="n">columns</span>
    <span class="n">cell_list</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">sc_adata</span><span class="o">.</span><span class="n">obs_names</span><span class="p">)</span>
    <span class="c1">#normalization</span>
    <span class="n">bulk_adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s1">&#39;mapping_ori&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">bulk_adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">normalization</span><span class="p">:</span>
        <span class="n">sc_adata</span><span class="o">=</span><span class="n">utils</span><span class="o">.</span><span class="n">normalization_cpm</span><span class="p">(</span><span class="n">sc_adata</span><span class="p">,</span><span class="n">scale_factors</span><span class="o">=</span><span class="mi">100000</span><span class="p">,</span><span class="n">trans_method</span><span class="o">=</span><span class="s2">&quot;log&quot;</span><span class="p">)</span>
        <span class="n">bulk_adata</span><span class="o">=</span><span class="n">utils</span><span class="o">.</span><span class="n">normalization_cpm</span><span class="p">(</span><span class="n">bulk_adata</span><span class="p">,</span><span class="n">scale_factors</span><span class="o">=</span><span class="mi">100000</span><span class="p">,</span><span class="n">trans_method</span><span class="o">=</span><span class="s2">&quot;log&quot;</span><span class="p">)</span>
    <span class="n">bulk_adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s1">&#39;mapping_nor&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">bulk_adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">input_sc_data</span> <span class="o">=</span> <span class="n">get</span><span class="o">.</span><span class="n">count_data</span><span class="p">(</span><span class="n">sc_adata</span><span class="p">,</span><span class="n">counts_location</span><span class="o">=</span><span class="n">sc_layer</span><span class="p">)</span>
    <span class="n">bulk_data</span> <span class="o">=</span> <span class="n">get</span><span class="o">.</span><span class="n">count_data</span><span class="p">(</span><span class="n">bulk_adata</span><span class="p">,</span><span class="n">counts_location</span><span class="o">=</span><span class="n">bulk_layer</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">reorder</span><span class="p">:</span>
        <span class="n">intersect_gene</span> <span class="o">=</span> <span class="n">input_sc_data</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">bulk_data</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
        <span class="n">input_sc_data</span> <span class="o">=</span> <span class="n">input_sc_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">intersect_gene</span><span class="p">,:]</span>
        <span class="n">bulk_data</span> <span class="o">=</span> <span class="n">bulk_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">intersect_gene</span><span class="p">,:]</span>

    <span class="n">sc_data</span> <span class="o">=</span> <span class="n">utils</span><span class="o">.</span><span class="n">normal_center</span><span class="p">(</span><span class="n">input_sc_data</span><span class="p">)</span>
    <span class="n">bulk_data</span> <span class="o">=</span> <span class="n">utils</span><span class="o">.</span><span class="n">normal_center</span><span class="p">(</span><span class="n">bulk_data</span><span class="p">)</span>
    <span class="n">bulk_adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s1">&#39;normal_center&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">bulk_data</span><span class="o">.</span><span class="n">T</span><span class="o">.</span><span class="n">values</span>
    <span class="n">sample</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">cell_prop</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="n">sc_data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>
    <span class="n">mapped_cor</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">sample_ori</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">cell_prop</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="n">sc_data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>
    <span class="n">sc_mapping_dict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">([(</span><span class="n">k</span><span class="p">,[])</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">bulk_data</span><span class="o">.</span><span class="n">columns</span><span class="p">])</span>
    <span class="k">if</span> <span class="n">multiprocessing</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">cpu_count</span><span class="p">()</span><span class="o">&lt;</span><span class="mi">2</span><span class="p">:</span>
            <span class="n">cpu_num</span> <span class="o">=</span> <span class="n">cpu_count</span><span class="p">()</span>
        <span class="c1"># compute person correlation and select sc according to person correlation.</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;multiprocessing mode, cpu count is </span><span class="si">{</span><span class="n">cpu_num</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">with</span> <span class="n">Pool</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">cpu_num</span><span class="p">))</span> <span class="k">as</span> <span class="n">p</span><span class="p">:</span>
            <span class="n">results</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">starmap</span><span class="p">(</span><span class="n">_bulk_mapping_parallel</span><span class="p">,</span> <span class="p">[(</span><span class="n">i</span><span class="p">,</span> <span class="n">cell_num</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">i</span><span class="p">,:],</span> <span class="n">bulk_data</span><span class="p">,</span> <span class="n">sc_data</span><span class="p">,</span> <span class="n">cell_list</span><span class="p">,</span> <span class="n">meta_dict</span><span class="p">,</span> <span class="n">cellname_list</span><span class="p">,</span> <span class="n">input_sc_data</span><span class="p">)</span>
             <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">cell_num</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])])</span>
            <span class="nb">print</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
        <span class="c1"># postprocessing</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">sample_ori_i</span><span class="p">,</span><span class="n">sample_i</span><span class="p">,</span> <span class="n">mapped_cor_i</span><span class="p">,</span> <span class="n">sc_mapping_dict_i</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">results</span><span class="p">):</span>
            <span class="n">sample_ori</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">sample_ori_i</span><span class="p">)</span>
            <span class="n">sample</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">sample_i</span><span class="p">)</span>
            <span class="n">mapped_cor</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mapped_cor_i</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">sc_mapping_dict_i</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                <span class="n">sc_mapping_dict</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">sc_mapping_dict_i</span><span class="p">[</span><span class="n">k</span><span class="p">])</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">index_num</span><span class="p">,</span> <span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">sample_num</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">tqdm</span><span class="p">(</span><span class="n">cell_num</span><span class="o">.</span><span class="n">iterrows</span><span class="p">())):</span>
            <span class="n">sample_cor</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">bulk_data</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="n">bulk_data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span><span class="n">sc_data</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
            <span class="n">cor_index</span> <span class="o">=</span> <span class="n">cell_list</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">sample_cor</span><span class="p">)]</span>
            <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">cellname</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">cellname_list</span><span class="p">):</span>
                <span class="n">mask</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">cor_index</span><span class="p">,</span> <span class="n">meta_dict</span><span class="p">[</span><span class="n">cellname</span><span class="p">])</span>
                <span class="n">sub_cell</span> <span class="o">=</span> <span class="n">cor_index</span><span class="p">[</span><span class="n">mask</span><span class="p">]</span>
                <span class="n">sub_cell</span> <span class="o">=</span> <span class="n">sub_cell</span><span class="p">[:</span><span class="nb">int</span><span class="p">(</span><span class="n">sample_num</span><span class="p">[</span><span class="n">j</span><span class="p">])]</span>
                <span class="n">sc_mapping_dict</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">sub_cell</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;sample </span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2"> done.&quot;</span><span class="p">)</span>
            <span class="n">sample_ori</span><span class="p">[</span><span class="n">index_num</span><span class="p">,:]</span> <span class="o">=</span> <span class="n">input_sc_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span><span class="n">sc_mapping_dict</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
            <span class="n">sample</span><span class="p">[</span><span class="n">index_num</span><span class="p">,:]</span> <span class="o">=</span> <span class="n">sc_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span><span class="n">sc_mapping_dict</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
            <span class="n">mapped_cor_i</span> <span class="o">=</span> <span class="n">utils</span><span class="o">.</span><span class="n">pear</span><span class="p">(</span><span class="n">sample</span><span class="p">[</span><span class="n">index_num</span><span class="p">,:],</span><span class="n">bulk_data</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">)</span><span class="o">.</span><span class="n">item</span><span class="p">()</span>
            <span class="n">mapped_cor</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mapped_cor_i</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;initial mapping solution:&#39;</span><span class="p">,</span><span class="s2">&quot;min correlation&quot;</span><span class="p">,</span> <span class="nb">min</span><span class="p">(</span><span class="n">mapped_cor</span><span class="p">),</span><span class="s2">&quot;average correlation&quot;</span><span class="p">,</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">mapped_cor</span><span class="p">),</span><span class="s2">&quot;max correlation&quot;</span><span class="p">,</span> <span class="nb">max</span><span class="p">(</span><span class="n">mapped_cor</span><span class="p">))</span>
    <span class="n">bulk_adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;cell_number&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">cell_matrix</span><span class="p">,</span><span class="n">index</span><span class="o">=</span><span class="n">cell_prop</span><span class="o">.</span><span class="n">index</span><span class="p">,</span><span class="n">columns</span><span class="o">=</span><span class="n">cell_prop</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
    <span class="n">bulk_adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s1">&#39;mapping&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">sample</span><span class="o">/</span><span class="n">n_cell</span>
    <span class="n">bulk_adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s1">&#39;mapping_ori&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">sample_ori</span><span class="o">/</span><span class="n">n_cell</span>
    <span class="k">if</span> <span class="n">filter_gene</span><span class="p">:</span>
        <span class="kn">from</span> <span class="nn">sklearn.metrics.pairwise</span> <span class="kn">import</span> <span class="n">cosine_similarity</span>
        <span class="n">gene_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">similarity_list</span><span class="o">=</span><span class="p">[]</span>
        <span class="n">data_ori</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">bulk_adata</span><span class="o">.</span><span class="n">X</span><span class="p">,</span><span class="n">index</span><span class="o">=</span><span class="n">bulk_adata</span><span class="o">.</span><span class="n">obs_names</span><span class="p">,</span><span class="n">columns</span><span class="o">=</span><span class="n">bulk_adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">)</span>
        <span class="n">data_mapping</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">bulk_adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s1">&#39;mapping_ori&#39;</span><span class="p">],</span><span class="n">index</span><span class="o">=</span><span class="n">bulk_adata</span><span class="o">.</span><span class="n">obs_names</span><span class="p">,</span><span class="n">columns</span><span class="o">=</span><span class="n">bulk_adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">bulk_adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">:</span>
            <span class="n">similarity</span> <span class="o">=</span> <span class="n">cosine_similarity</span><span class="p">(</span><span class="n">data_ori</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">),</span> <span class="n">data_mapping</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">))</span>
            <span class="k">if</span> <span class="n">similarity</span> <span class="o">&gt;</span> <span class="mf">0.6</span><span class="p">:</span>
                <span class="n">similarity_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">similarity</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
                <span class="n">gene_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
        <span class="n">bulk_adata</span><span class="o">=</span><span class="n">bulk_adata</span><span class="p">[:,</span><span class="n">gene_list</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Gene cosin similarity:&#39;</span><span class="p">,</span><span class="s2">&quot;min value&quot;</span><span class="p">,</span> <span class="nb">min</span><span class="p">(</span><span class="n">similarity_list</span><span class="p">),</span><span class="s2">&quot;average value&quot;</span><span class="p">,</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">similarity_list</span><span class="p">),</span><span class="s2">&quot;max value&quot;</span><span class="p">,</span> <span class="nb">max</span><span class="p">(</span><span class="n">similarity_list</span><span class="p">))</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;The number of reconstructed gene:</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">gene_list</span><span class="p">)</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>  
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Time to finish mapping: </span><span class="si">{</span><span class="nb">round</span><span class="p">(</span><span class="n">time</span><span class="o">.</span><span class="n">perf_counter</span><span class="p">()</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">start_t</span><span class="p">,</span><span class="w"> </span><span class="mi">2</span><span class="p">)</span><span class="si">}</span><span class="s1"> seconds&#39;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;=========================================================================================================================================&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">save</span><span class="p">:</span>
        <span class="n">out_dir</span> <span class="o">=</span> <span class="n">utils</span><span class="o">.</span><span class="n">check_paths</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">out_dir</span><span class="si">}</span><span class="s1">/output&#39;</span><span class="p">)</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">([(</span><span class="n">k</span><span class="p">,</span> <span class="n">v</span><span class="p">)</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">lst</span> <span class="ow">in</span> <span class="n">sc_mapping_dict</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">lst</span><span class="p">],</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;sample_id&#39;</span><span class="p">,</span> <span class="s1">&#39;cell_id&#39;</span><span class="p">])</span>
        <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">out_dir</span><span class="si">}</span><span class="s2">/bulk_data_</span><span class="si">{</span><span class="n">project</span><span class="si">}</span><span class="s2">_mapping.csv&quot;</span><span class="p">)</span>
        <span class="n">bulk_adata</span><span class="o">.</span><span class="n">write_h5ad</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">out_dir</span><span class="si">}</span><span class="s2">/bulk_data_</span><span class="si">{</span><span class="n">project</span><span class="si">}</span><span class="s2">_mapping.h5ad&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">df</span><span class="p">,</span><span class="n">bulk_adata</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div><hr />
<p><div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">st_mapping</span><span class="p">(</span><span class="n">st_adata</span><span class="p">,</span>
               <span class="n">sc_adata</span><span class="p">,</span>
               <span class="n">out_dir</span><span class="p">,</span>
               <span class="n">project</span><span class="p">,</span>
               <span class="n">annotation_key</span><span class="p">,</span>
               <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
</code></pre></div>
</p>


<div class="doc doc-object doc-function">



<a id="cytobulk.tools._mapping.st_mapping"></a>
    <div class="doc doc-contents first">

        <p>Run spatial transcriptomics mapping with single-cell RNA-seq data.</p>
<p>This function maps spatial transcriptomics (ST) data to single-cell RNA-seq (scRNA-seq) data. It aligns cell type compositions and estimates spatial distributions.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>st_adata</code>
            </td>
            <td>
                  <code><span title="anndata.AnnData">AnnData</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>An :class:<code>~anndata.AnnData</code> object containing spatial transcriptomics data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>sc_adata</code>
            </td>
            <td>
                  <code><span title="anndata.AnnData">AnnData</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>An :class:<code>~anndata.AnnData</code> object containing single-cell RNA-seq data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>seed</code>
            </td>
            <td>
                  <code>int, optional (default: 0)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Seed for random number generation to ensure reproducibility.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>annotation_key</code>
            </td>
            <td>
                  <code>string, optional (default: &#39;celltype_minor&#39;)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Key in <code>sc_adata</code> for cell type annotations.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>sc_downsample</code>
            </td>
            <td>
                  <code>bool, optional (default: False)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to downsample scRNA-seq data to a maximum number of transcripts per cell.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>scRNA_max_transcripts_per_cell</code>
            </td>
            <td>
                  <code>int, optional (default: 1500)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Maximum number of transcripts per cell for downsampling.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>sampling_method</code>
            </td>
            <td>
                  <code>string, optional (default: &#39;duplicates&#39;)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Method for sampling single cells based on cell type composition.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>out_dir</code>
            </td>
            <td>
                  <code>string, optional (default: &#39;.&#39;)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Directory to save output files.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>project</code>
            </td>
            <td>
                  <code>string, optional (default: &#39;test&#39;)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Project name for output file naming.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>mean_cell_numbers</code>
            </td>
            <td>
                  <code>int, optional (default: 8)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Average number of cells per spot used for estimation.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>save_reconstructed_st</code>
            </td>
            <td>
                  <code>bool, optional (default: True)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to save the reconstructed spatial transcriptomics data.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>reconstructed_sc</code></td>            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>DataFrame containing the mapping of single-cell IDs to spatial spot IDs.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>cytobulk\tools\_mapping.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">st_mapping</span><span class="p">(</span><span class="n">st_adata</span><span class="p">,</span>
               <span class="n">sc_adata</span><span class="p">,</span>
               <span class="n">out_dir</span><span class="p">,</span>
               <span class="n">project</span><span class="p">,</span>
               <span class="n">annotation_key</span><span class="p">,</span>
               <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Run spatial transcriptomics mapping with single-cell RNA-seq data.</span>

<span class="sd">    This function maps spatial transcriptomics (ST) data to single-cell RNA-seq (scRNA-seq) data. It aligns cell type compositions and estimates spatial distributions.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    st_adata : anndata.AnnData</span>
<span class="sd">        An :class:`~anndata.AnnData` object containing spatial transcriptomics data.</span>

<span class="sd">    sc_adata : anndata.AnnData</span>
<span class="sd">        An :class:`~anndata.AnnData` object containing single-cell RNA-seq data.</span>

<span class="sd">    seed : int, optional (default: 0)</span>
<span class="sd">        Seed for random number generation to ensure reproducibility.</span>

<span class="sd">    annotation_key : string, optional (default: &#39;celltype_minor&#39;)</span>
<span class="sd">        Key in `sc_adata` for cell type annotations.</span>

<span class="sd">    sc_downsample : bool, optional (default: False)</span>
<span class="sd">        Whether to downsample scRNA-seq data to a maximum number of transcripts per cell.</span>

<span class="sd">    scRNA_max_transcripts_per_cell : int, optional (default: 1500)</span>
<span class="sd">        Maximum number of transcripts per cell for downsampling.</span>

<span class="sd">    sampling_method : string, optional (default: &#39;duplicates&#39;)</span>
<span class="sd">        Method for sampling single cells based on cell type composition.</span>

<span class="sd">    out_dir : string, optional (default: &#39;.&#39;)</span>
<span class="sd">        Directory to save output files.</span>

<span class="sd">    project : string, optional (default: &#39;test&#39;)</span>
<span class="sd">        Project name for output file naming.</span>

<span class="sd">    mean_cell_numbers : int, optional (default: 8)</span>
<span class="sd">        Average number of cells per spot used for estimation.</span>

<span class="sd">    save_reconstructed_st : bool, optional (default: True)</span>
<span class="sd">        Whether to save the reconstructed spatial transcriptomics data.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    reconstructed_sc : pandas.DataFrame</span>
<span class="sd">        DataFrame containing the mapping of single-cell IDs to spatial spot IDs.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">start_t</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">perf_counter</span><span class="p">()</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;=================================================================================================&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Start to mapping bulk data with single cell dataset.&#39;</span><span class="p">)</span>
    <span class="n">reconstructed_sc</span><span class="o">=</span><span class="n">_run_st_mapping</span><span class="p">(</span><span class="n">st_adata</span> <span class="o">=</span> <span class="n">st_adata</span><span class="p">,</span>
                                     <span class="n">sc_adata</span> <span class="o">=</span> <span class="n">sc_adata</span><span class="p">,</span>
                                     <span class="n">out_dir</span> <span class="o">=</span> <span class="n">out_dir</span><span class="p">,</span>
                                     <span class="n">project</span> <span class="o">=</span> <span class="n">project</span><span class="p">,</span>
                                     <span class="n">annotation_key</span><span class="o">=</span><span class="n">annotation_key</span><span class="p">,</span>
                                     <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Time to finish mapping: </span><span class="si">{</span><span class="nb">round</span><span class="p">(</span><span class="n">time</span><span class="o">.</span><span class="n">perf_counter</span><span class="p">()</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">start_t</span><span class="p">,</span><span class="w"> </span><span class="mi">2</span><span class="p">)</span><span class="si">}</span><span class="s1"> seconds&#39;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;=========================================================================================================================================&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">reconstructed_sc</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div><hr />
<p><div class="highlight"><pre><span></span><code><span class="k">def</span> <span class="nf">he_mapping</span><span class="p">(</span><span class="n">image_dir</span><span class="p">,</span>
               <span class="n">out_dir</span><span class="p">,</span>
               <span class="n">project</span><span class="p">,</span>
               <span class="n">lr_data</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
               <span class="n">sc_adata</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
               <span class="n">annotation_key</span><span class="o">=</span><span class="s2">&quot;curated_celltype&quot;</span><span class="p">,</span>
               <span class="n">k_neighbor</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span>
               <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>
               <span class="n">mapping_sc</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
               <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
</code></pre></div>
</p>


<div class="doc doc-object doc-function">



<a id="cytobulk.tools._mapping.he_mapping"></a>
    <div class="doc doc-contents first">

        <p>Run H&amp;E-stained image cell type mapping with single-cell RNA-seq data.</p>
<p>This function predicts cell types from H&amp;E-stained histology images and aligns them with single-cell RNA-seq (scRNA-seq) data using optimal transport. It computes spatial distributions and matches cell types between the image and single-cell data.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>image_dir</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the directory containing H&amp;E-stained images.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>out_dir</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Directory where the output files will be saved.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>project</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Name of the project, used for naming output files.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>lr_data</code>
            </td>
            <td>
                  <code>pandas.DataFrame, optional (default: None)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A DataFrame containing ligand-receptor pair data with columns 'ligand' and 'receptor'.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>sc_adata</code>
            </td>
            <td>
                  <code>anndata.AnnData, optional (default: None)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>An :class:<code>~anndata.AnnData</code> object containing single-cell RNA-seq data with gene expression profiles.</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>annotation_key</code>
            </td>
            <td>
                  <code>str, optional (default: &#34;curated_celltype&#34;)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Key in <code>sc_adata.obs</code> for cell type annotations.</p>
              </div>
            </td>
            <td>
                  <code>&#39;curated_celltype&#39;</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>k_neighbor</code>
            </td>
            <td>
                  <code>int, optional (default: 30)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of neighbors to consider when constructing the graph for H&amp;E image data.</p>
              </div>
            </td>
            <td>
                  <code>30</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>alpha</code>
            </td>
            <td>
                  <code>float, optional (default: 0.5)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Trade-off parameter for the Fused Gromov-Wasserstein optimal transport, controlling the balance between graph structure and feature matching (value between 0 and 1).</p>
              </div>
            </td>
            <td>
                  <code>0.5</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>mapping_sc</code>
            </td>
            <td>
                  <code>bool, optional (default: True)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to perform mapping between H&amp;E image cell data and single-cell RNA-seq data. If False, only H&amp;E image cell type predictions are returned.</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>**kwargs</code>
            </td>
            <td>
                  <code>dict</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Additional arguments (not used in this implementation).</p>
              </div>
            </td>
            <td>
                  <code>{}</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>cell_coordinates</code></td>            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>DataFrame containing cell coordinates and their predicted cell types from H&amp;E-stained images.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>df</code></td>            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>DataFrame containing matching results between H&amp;E image cells and single-cell data, including spatial coordinates, cell types, and matched single-cell IDs.</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>filtered_adata</code></td>            <td>
                  <code><span title="anndata.AnnData">AnnData</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A filtered :class:<code>~anndata.AnnData</code> object containing only the single-cell data that matches with cells from H&amp;E-stained images.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>cytobulk\tools\_mapping.py</code></summary>
              <div class="codehilite"><table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">he_mapping</span><span class="p">(</span><span class="n">image_dir</span><span class="p">,</span>
               <span class="n">out_dir</span><span class="p">,</span>
               <span class="n">project</span><span class="p">,</span>
               <span class="n">lr_data</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
               <span class="n">sc_adata</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
               <span class="n">annotation_key</span><span class="o">=</span><span class="s2">&quot;curated_celltype&quot;</span><span class="p">,</span>
               <span class="n">k_neighbor</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span>
               <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>
               <span class="n">mapping_sc</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
               <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>

<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Run H&amp;E-stained image cell type mapping with single-cell RNA-seq data.</span>

<span class="sd">    This function predicts cell types from H&amp;E-stained histology images and aligns them with single-cell RNA-seq (scRNA-seq) data using optimal transport. It computes spatial distributions and matches cell types between the image and single-cell data.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    image_dir : str</span>
<span class="sd">        Path to the directory containing H&amp;E-stained images.</span>

<span class="sd">    out_dir : str</span>
<span class="sd">        Directory where the output files will be saved.</span>

<span class="sd">    project : str</span>
<span class="sd">        Name of the project, used for naming output files.</span>

<span class="sd">    lr_data : pandas.DataFrame, optional (default: None)</span>
<span class="sd">        A DataFrame containing ligand-receptor pair data with columns &#39;ligand&#39; and &#39;receptor&#39;.</span>

<span class="sd">    sc_adata : anndata.AnnData, optional (default: None)</span>
<span class="sd">        An :class:`~anndata.AnnData` object containing single-cell RNA-seq data with gene expression profiles.</span>

<span class="sd">    annotation_key : str, optional (default: &quot;curated_celltype&quot;)</span>
<span class="sd">        Key in `sc_adata.obs` for cell type annotations.</span>

<span class="sd">    k_neighbor : int, optional (default: 30)</span>
<span class="sd">        Number of neighbors to consider when constructing the graph for H&amp;E image data.</span>

<span class="sd">    alpha : float, optional (default: 0.5)</span>
<span class="sd">        Trade-off parameter for the Fused Gromov-Wasserstein optimal transport, controlling the balance between graph structure and feature matching (value between 0 and 1).</span>

<span class="sd">    mapping_sc : bool, optional (default: True)</span>
<span class="sd">        Whether to perform mapping between H&amp;E image cell data and single-cell RNA-seq data. If False, only H&amp;E image cell type predictions are returned.</span>

<span class="sd">    **kwargs : dict</span>
<span class="sd">        Additional arguments (not used in this implementation).</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    cell_coordinates : pandas.DataFrame</span>
<span class="sd">        DataFrame containing cell coordinates and their predicted cell types from H&amp;E-stained images.</span>

<span class="sd">    df : pandas.DataFrame</span>
<span class="sd">        DataFrame containing matching results between H&amp;E image cells and single-cell data, including spatial coordinates, cell types, and matched single-cell IDs.</span>

<span class="sd">    filtered_adata : anndata.AnnData</span>
<span class="sd">        A filtered :class:`~anndata.AnnData` object containing only the single-cell data that matches with cells from H&amp;E-stained images.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">start_t</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">perf_counter</span><span class="p">()</span>
    <span class="n">file_dir</span> <span class="o">=</span> <span class="n">resource_filename</span><span class="p">(</span><span class="vm">__name__</span><span class="p">,</span> <span class="s1">&#39;model/pretrained_models/&#39;</span><span class="p">)</span>
    <span class="n">file_name</span> <span class="o">=</span> <span class="s1">&#39;DeepCMorph_Datasets_Combined_41_classes_acc_8159.pth&#39;</span>

    <span class="c1"># The download URL for the file (replace with the actual URL)</span>
    <span class="n">download_url</span> <span class="o">=</span> <span class="s2">&quot;https://drive.google.com/file/d/1M5dNB5XtrboQcxuI59rPNdR29nLe6UFC/view?usp=sharing&quot;</span>

    <span class="c1"># Ensure the file exists; if not, download it</span>
    <span class="n">get</span><span class="o">.</span><span class="n">ensure_file_exists</span><span class="p">(</span><span class="n">file_dir</span><span class="p">,</span> <span class="n">file_name</span><span class="p">,</span> <span class="n">download_url</span><span class="p">)</span>
    <span class="n">cell_coordinates</span> <span class="o">=</span> <span class="n">inference_cell_type_from_he_image</span><span class="p">(</span><span class="n">image_dir</span><span class="p">,</span>
                                                         <span class="n">out_dir</span><span class="p">,</span>
                                                         <span class="n">project</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">mapping_sc</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;preprocessing of single cell data&quot;</span><span class="p">)</span>
        <span class="n">lr_genes</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">((</span><span class="n">lr_data</span><span class="p">[</span><span class="s1">&#39;ligand&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">lr_data</span><span class="p">[</span><span class="s1">&#39;receptor&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">)))</span>
        <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_cells</span><span class="p">(</span><span class="n">sc_adata</span><span class="p">,</span> <span class="n">min_genes</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>  <span class="c1"># filter</span>
        <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes</span><span class="p">(</span><span class="n">sc_adata</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>   <span class="c1"># filter</span>
        <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">sc_adata</span><span class="p">,</span> <span class="n">target_sum</span><span class="o">=</span><span class="mf">1e4</span><span class="p">)</span>  <span class="c1"># nor</span>
        <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">sc_adata</span><span class="p">)</span>  <span class="c1"># log</span>
        <span class="n">common_gene</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">intersect1d</span><span class="p">(</span><span class="n">lr_genes</span><span class="p">,</span> <span class="n">sc_adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">)</span>
        <span class="n">sc_adata</span> <span class="o">=</span> <span class="n">sc_adata</span><span class="p">[:,</span> <span class="n">common_gene</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">lr_data</span> <span class="o">=</span> <span class="n">lr_data</span><span class="p">[</span>
            <span class="p">(</span><span class="n">lr_data</span><span class="p">[</span><span class="s1">&#39;ligand&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">common_gene</span><span class="p">))</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">lr_data</span><span class="p">[</span><span class="s1">&#39;receptor&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">common_gene</span><span class="p">))</span>
        <span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="n">adata_cell_types</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">sc_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">annotation_key</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>
        <span class="n">coordinates_cell_types</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">cell_coordinates</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>


        <span class="c1"># common_celltype</span>
        <span class="n">common_cell_types</span> <span class="o">=</span> <span class="n">adata_cell_types</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">coordinates_cell_types</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Common cell types: </span><span class="si">{</span><span class="n">common_cell_types</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># filter adata and cell_coordinates</span>
        <span class="n">sc_adata</span> <span class="o">=</span> <span class="n">sc_adata</span><span class="p">[</span><span class="n">sc_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">annotation_key</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">common_cell_types</span><span class="p">),</span> <span class="p">:]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">cell_coordinates</span> <span class="o">=</span> <span class="n">cell_coordinates</span><span class="p">[</span><span class="n">cell_coordinates</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">common_cell_types</span><span class="p">)]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>


        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;loading graph for H&amp;E image...&quot;</span><span class="p">)</span>
        <span class="n">graph1_adj</span><span class="p">,</span> <span class="n">graph1_labels</span> <span class="o">=</span> <span class="n">load_graph1</span><span class="p">(</span><span class="n">cell_coordinates</span><span class="p">,</span><span class="n">k</span><span class="o">=</span><span class="n">k_neighbor</span><span class="p">)</span>

        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;loading graph for single cell data with LR affinity...&quot;</span><span class="p">)</span>
        <span class="n">graph2_dist</span><span class="p">,</span> <span class="n">graph2_labels</span><span class="p">,</span><span class="n">sc_adata</span> <span class="o">=</span> <span class="n">load_graph2_with_LR_affinity</span><span class="p">(</span><span class="n">sc_adata</span><span class="p">,</span> 
                                                                <span class="n">graph1_labels</span><span class="p">,</span>
                                                                <span class="n">lr_data</span><span class="p">,</span>
                                                                <span class="n">annotation_key</span><span class="p">)</span>

        <span class="n">graph2_dist</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan_to_num</span><span class="p">(</span><span class="n">graph2_dist</span><span class="p">,</span> <span class="n">nan</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">nanmax</span><span class="p">(</span><span class="n">graph2_dist</span><span class="p">),</span> <span class="n">posinf</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">nanmax</span><span class="p">(</span><span class="n">graph2_dist</span><span class="p">),</span> <span class="n">neginf</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;compute cost matrix&quot;</span><span class="p">)</span>
        <span class="n">cost_matrix</span> <span class="o">=</span> <span class="n">construct_cost_matrix</span><span class="p">(</span><span class="n">graph1_labels</span><span class="p">,</span> <span class="n">graph2_labels</span><span class="p">)</span>

        <span class="n">cost_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan_to_num</span><span class="p">(</span><span class="n">cost_matrix</span><span class="p">,</span> <span class="n">nan</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">nanmax</span><span class="p">(</span><span class="n">cost_matrix</span><span class="p">),</span> <span class="n">posinf</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">nanmax</span><span class="p">(</span><span class="n">cost_matrix</span><span class="p">),</span> <span class="n">neginf</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;optimal transport...&quot;</span><span class="p">)</span>
        <span class="n">p</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="n">graph1_adj</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">/</span> <span class="n">graph1_adj</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">q</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="n">graph2_dist</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">/</span> <span class="n">graph2_dist</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

        <span class="n">p</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan_to_num</span><span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="n">nan</span><span class="o">=</span><span class="mf">1.0</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">p</span><span class="p">),</span> <span class="n">posinf</span><span class="o">=</span><span class="mf">1.0</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">p</span><span class="p">),</span> <span class="n">neginf</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
        <span class="n">q</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan_to_num</span><span class="p">(</span><span class="n">q</span><span class="p">,</span> <span class="n">nan</span><span class="o">=</span><span class="mf">1.0</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">q</span><span class="p">),</span> <span class="n">posinf</span><span class="o">=</span><span class="mf">1.0</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">q</span><span class="p">),</span> <span class="n">neginf</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

        <span class="c1"># Fused Gromov-Wasserstein</span>
<span class="w">        </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        gw_trans, log = ot.gromov.BAPG_fused_gromov_wasserstein(</span>
<span class="sd">            cost_matrix, graph1_adj, graph2_dist, p, q, alpha=alpha, log=True</span>
<span class="sd">        )</span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="n">ot_plan</span> <span class="o">=</span> <span class="n">ot</span><span class="o">.</span><span class="n">gromov</span><span class="o">.</span><span class="n">fused_gromov_wasserstein</span><span class="p">(</span>
        <span class="n">cost_matrix</span><span class="p">,</span> <span class="n">graph1_adj</span><span class="p">,</span> <span class="n">graph2_dist</span><span class="p">,</span> <span class="n">p</span><span class="p">,</span> <span class="n">q</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="n">alpha</span><span class="p">,</span> <span class="n">loss_fun</span><span class="o">=</span><span class="s1">&#39;square_loss&#39;</span>
        <span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Time to finish mapping: </span><span class="si">{</span><span class="nb">round</span><span class="p">(</span><span class="n">time</span><span class="o">.</span><span class="n">perf_counter</span><span class="p">()</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">start_t</span><span class="p">,</span><span class="w"> </span><span class="mi">2</span><span class="p">)</span><span class="si">}</span><span class="s1"> seconds&#39;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;=========================================================================================================================================&quot;</span><span class="p">)</span>
        <span class="c1"># Step 5: matching file</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;build matching file...&quot;</span><span class="p">)</span>
        <span class="n">locations</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">graph1_adj</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span> 
        <span class="n">cells</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">graph2_dist</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span> 
        <span class="n">matches</span> <span class="o">=</span> <span class="n">extract_matching_relationships</span><span class="p">(</span><span class="n">ot_plan</span><span class="p">,</span> <span class="n">locations</span><span class="p">,</span> <span class="n">cells</span><span class="p">)</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">matches</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;location&quot;</span><span class="p">,</span> <span class="s2">&quot;cell&quot;</span><span class="p">])</span>
        <span class="n">location_mapping</span> <span class="o">=</span> <span class="n">cell_coordinates</span><span class="p">[[</span><span class="s2">&quot;x&quot;</span><span class="p">,</span> <span class="s2">&quot;y&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">to_dict</span><span class="p">(</span><span class="n">orient</span><span class="o">=</span><span class="s2">&quot;index&quot;</span><span class="p">)</span>
        <span class="n">df</span><span class="p">[</span><span class="s2">&quot;x&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;location&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">c</span><span class="p">:</span> <span class="n">cell_coordinates</span><span class="p">[</span><span class="s1">&#39;x&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="n">c</span><span class="p">])</span>
        <span class="n">df</span><span class="p">[</span><span class="s2">&quot;y&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;location&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">c</span><span class="p">:</span> <span class="n">cell_coordinates</span><span class="p">[</span><span class="s1">&#39;y&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="n">c</span><span class="p">])</span>
        <span class="n">df</span><span class="p">[</span><span class="s2">&quot;he_cell_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;location&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">c</span><span class="p">:</span> <span class="n">cell_coordinates</span><span class="p">[</span><span class="s1">&#39;cell_type&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="n">c</span><span class="p">])</span>
        <span class="n">df</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;cell&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">c</span><span class="p">:</span> <span class="n">graph2_labels</span><span class="p">[</span><span class="n">c</span><span class="p">])</span>
        <span class="n">df</span><span class="p">[</span><span class="s2">&quot;cell_id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;cell&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">c</span><span class="p">:</span> <span class="n">sc_adata</span><span class="o">.</span><span class="n">obs_names</span><span class="p">[</span><span class="n">c</span><span class="p">])</span>
        <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">out_dir</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">project</span><span class="si">}</span><span class="s2">_matching_results.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">df_cell_ids</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;cell_id&#39;</span><span class="p">]</span>
        <span class="n">adata_cell_ids</span> <span class="o">=</span> <span class="n">sc_adata</span><span class="o">.</span><span class="n">obs_names</span> 
        <span class="n">common_cell_ids</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">df_cell_ids</span><span class="p">)</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">adata_cell_ids</span><span class="p">))</span>
        <span class="n">filtered_df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;cell_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">common_cell_ids</span><span class="p">)]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">filtered_adata</span> <span class="o">=</span> <span class="n">sc_adata</span><span class="p">[</span><span class="n">sc_adata</span><span class="o">.</span><span class="n">obs_names</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">common_cell_ids</span><span class="p">)]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">filtered_df</span> <span class="o">=</span> <span class="n">filtered_df</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;cell_id&#39;</span><span class="p">)</span>
        <span class="c1">#filtered_adata.obs = pd.DataFrame()</span>
        <span class="n">filtered_adata</span><span class="o">.</span><span class="n">obs</span> <span class="o">=</span> <span class="n">filtered_df</span>
        <span class="n">filtered_adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="s2">&quot;gene&quot;</span>
        <span class="n">filtered_adata</span><span class="o">.</span><span class="n">write_h5ad</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">out_dir</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">project</span><span class="si">}</span><span class="s2">_matching_adata.h5ad&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">cell_coordinates</span><span class="p">,</span><span class="n">df</span><span class="p">,</span><span class="n">filtered_adata</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">cell_coordinates</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    
      
      <nav class="md-footer__inner md-grid" aria-label="Footer" >
        
          
          <a href="../_deconv/" class="md-footer__link md-footer__link--prev" aria-label="Previous: deconv">
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
            </div>
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Previous
              </span>
              <div class="md-ellipsis">
                deconv
              </div>
            </div>
          </a>
        
        
          
          <a href="../_segmentation/" class="md-footer__link md-footer__link--next" aria-label="Next: segmentation">
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Next
              </span>
              <div class="md-ellipsis">
                segmentation
              </div>
            </div>
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11z"/></svg>
            </div>
          </a>
        
      </nav>
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Copyright &copy; 2023 WANG Xueying WANG Yian
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["search.suggest", "search.highlight", "search.share\\", "navigation.footer"], "search": "../../assets/javascripts/workers/search.6ce7567c.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.88dd0f4e.min.js"></script>
      
    
  </body>
</html>